---
title: "Alpha diversity"
author: "Marcela Aragon"
date: "29/07/2024"
output: html_document
---

This script will deal with calculating alpha diversity for 16S and ITS 

#Loading packages 

```{r}

#Load libraries

#microbiome data
library(phyloseq)
library(microbiome)
library(metagMisc)
library(vegan)
library(ape)

#data handling
library(dplyr)
library(tidyr)
library(tibble)
library(readxl)

#stats
library(picante) #faith's Phylogenetic Diversity
library(stats) #stats
library(DHARMa) #check residuals
library(lme4) #run glms
library(fitdistrplus) #checks data distribution
library(emmeans)
library(rempsyc) #tables
library(magrittr) #pipe with emmeans
library(effectsize)
library(scales)#rounds p value

#ploting
library(ggplot2)
library(viridis)
library(GGally)
library(ggpubr)


```

## Settings for plots
```{r}

#Set plots theme
ggplot2::theme_set(theme_bw())

# Set default axis looks
axis_looks <- theme(axis.text.x = element_text(colour = "black", size = 9,
                                               face = "bold", angle=0, hjust=0.5))+ #text in X
  theme(axis.text.y = element_text(colour = "black", size = 9, face = "bold"))+ #text in y
  theme(axis.title=element_text(size=11, face = "bold"))+ #all texts
  theme(axis.title.y = element_text(margin = ggplot2::margin(t = 0, r = 10, b = 0, l = 0)))+
  theme(panel.grid.major = element_blank(), #no grid
        panel.grid.minor = element_blank()) #no grid

#Set colors
color_stress <- c("#BFBFBF", "#E69F00", "#319B67")
color_experiment <- c("#CC3399", "#70AD47")

```

#ITS

##Loading data
```{r}
#ITS
load(file = "./amplicon_data/Phyloseq_objects/ps_ITS.RData")
ps_ITS

#let's quickly check how many replicates we have left per experiment
#and how is the library size amongst them
df_summ <- ps_ITS %>% 
    sample_data() %>% 
    group_by(Experiment, Stress)%>%
    summarise(Replicates = n(),
              library_size = mean(libsize_after_filtering),
              stdev_libsize =sd(libsize_after_filtering))%>%
    as.data.frame()

df_summ%>%
    ggplot(., aes(x=Stress,y=library_size, label=Replicates))+
    geom_bar(stat = "identity")+
    geom_text(size=8)+
    facet_wrap(~Experiment, scales = "free_x") #ok

#save this table
table <- nice_table(df_summ,
              title = c("Table X. Summary for ITS"))
flextable::save_as_docx(table, path = "./Tables/01_alpha_beta_diversity/summary_ITS.docx")


#get sample data df 
sample_data <- as.data.frame(sample_data(ps_ITS))

#split by experiment 
ITS_exp <- phyloseq_sep_variable(ps_ITS, variable="Experiment", drop_zeroes = T)
ITS_exp <- ITS_exp[-3] #let's remove EPG for now

#rarefy to even depth
#ITS_exp_rar <-lapply(ITS_exp, function(ps){
#  rarefy_even_depth(ps, sample.size = min(sample_sums(ps)),
#  rngseed = 123, replace = TRUE, trimOTUs = TRUE, verbose = TRUE)})

pITS_median <- lapply(ITS_exp, function(ps){
  total <- median(sample_sums(ps))
  standf <- function(x, t=total) round(t * (x / sum(x)))
  physeq_msd = transform_sample_counts(ps, standf)
  print((sample_sums(physeq_msd)))
  return(physeq_msd)
})


lapply(pITS_median, function(ps){ sample_sums(ps)}) 
#77,964 reads for Phyto
#71,489 reads for Insects

```

Phytohormones experiment has 6 replicates for controls, the rest is between 10-16 replicates per treatment.

##Alpha diversity
```{r}

#plot it
#p <- lapply(ITS_exp_rar, function(ps){plot_richness(ps, x= "plant_pathway_induced", scales="free_y")})

#lapply(p, function(plot){
#  plot + geom_boxplot(data=plot$data, aes(x= Experiment, y=value, color=NULL), alpha=0.1)+
#    scale_color_manual(values=color_experiment)})

#Observed(Richness):Total number of ASVs
#Chao1:Same as Observed but aims to calculate real number of ASVs using single- and doubletons
#Shannon(diversity index):Lower value, lower diversity
#Simpson(similarity index):Takes abundance (eveness) into account. Lower value, higher diversity


#get Observed, Shannon and Simpson indexes
richness <- lapply(pITS_median, function(ps){
  estimate_richness(ps,
            measures = c("Observed", "Shannon", "Simpson"))
})

#fix df 
richness <- lapply(richness, function(df){
  rownames(df) <- sub("X", "", rownames(df)) #removes X from rownames
  df$sample_name <- rownames(df) 
  return(df)
})

#Get Faith's phylogenetic diversity index
pd <- lapply(pITS_median, function(ps){
  pd(t(as.data.frame(ps@otu_table)), ps@phy_tree,include.root=T)
  })

#merge with metadata of ps object
richness$cond_phytohormones <- x<-cbind(as.data.frame(sample_data(pITS_median$cond_phytohormones)), richness$cond_phytohormones, pd$cond_phytohormones)

richness$cond_real_insects<- cbind(as.data.frame(sample_data(pITS_median$cond_real_insects)), richness$cond_real_insects, pd$cond_real_insects)

#select the columns needed
richness <- lapply(richness, function(df){
  df <- df %>% dplyr::select(c('Experiment',
                         'plant_pathway_induced',
                         'Amplicon_concentation_ngul',
                         'total_reads',
                         'libsize_after_filtering',
                         'shoot_biomass',"Observed",
                         "Shannon", "Simpson", "PD", "sample_name"))
})

#make another set with long format for plotting
richness_long <- lapply(richness, function(df){
  df_long <- reshape2::melt(df)})

#clean
rm(p)

```

For now, just work with Richness, Shannon and PD

###Plot

Plot it

```{r}


estimates <- c("Observed","Shannon", "PD")

#add info for y axis

limits <- data.frame(
  variable = c("Observed","Shannon","PD"),
  y_min = c(0, 0, 0), # Set your desired minimum y-axis limits
  y_max = c(200, 4, 40) # Set your desired maximum y-axis limits
)

#merge
richness_long_ITS_estimates <- lapply(richness_long, function(df){
  df <- filter(df, variable %in% estimates)
  df <- df %>% left_join(limits, by="variable")
  df$variable <- factor(df$variable, levels=c("Observed","Shannon","PD"))
  return(df)
  })


#Plot phytohormones
p.p <- ggplot(richness_long_ITS_estimates$cond_phytohormones,
       aes(x = plant_pathway_induced, y = value, fill=plant_pathway_induced))+
     geom_boxplot(aes(fill = plant_pathway_induced))+
     geom_jitter(data = richness_long_ITS_estimates$cond_phytohormones,
                 mapping = aes(x = plant_pathway_induced, y = value, fill=plant_pathway_induced),
              shape = 21, size = 1.5, color = "black", alpha=0.5,
              position = position_jitterdodge(jitter.height = .1, jitter.width = .5))+
     scale_fill_manual(values=color_stress)+
  geom_line() +
  geom_blank(aes(y = y_min)) + # Force the minimum y limit
  geom_blank(aes(y = y_max)) + # Force the maximum y limit
  facet_wrap(~ variable, scales = "free_y") +
  theme(legend.position = "none")+
    axis_looks

p.i <- ggplot(richness_long_ITS_estimates$cond_real_insects,
       aes(x = plant_pathway_induced, y = value, fill=plant_pathway_induced))+
     geom_boxplot(aes(fill = plant_pathway_induced))+
     geom_jitter(data = richness_long_ITS_estimates$cond_real_insects,
                 mapping = aes(x = plant_pathway_induced, y = value, fill=plant_pathway_induced),
              shape = 21, size = 1.5, color = "black", alpha=0.5,
              position = position_jitterdodge(jitter.height = .1, jitter.width = .5))+
     scale_fill_manual(values=color_stress)+
  geom_line() +
  geom_blank(aes(y = y_min)) + # Force the minimum y limit
  geom_blank(aes(y = y_max)) + # Force the maximum y limit
  facet_wrap(~ variable, scales = "free_y") +
  theme(legend.position = "none")+
    axis_looks


library(patchwork)

p.p / p.i

#save it
mapply(function(x)
  ggsave(
  plot = last_plot(),
  filename =x,
  path = "./Plots/01_alpha_beta_diversity",
  scale = 2.0,
  width = 1000,
  height = 800,
  units = "px",
  dpi = 300),
x=c("alpha_both_ITS.pdf","alpha_both_ITS.png"))


```


```{r}

estimates <- c("Observed","Shannon", "PD")

#Plot phytohormones
ggplot(subset(richness_long$cond_phytohormones, variable %in% estimates), aes(x = plant_pathway_induced, y = value, fill=plant_pathway_induced))+
     geom_boxplot(aes(fill = plant_pathway_induced))+
     geom_jitter(data = subset(richness_long$cond_phytohormones, variable %in% estimates),
                 mapping = aes(x = plant_pathway_induced, y = value, fill=plant_pathway_induced),
              shape = 21, size = 1.5, color = "black", alpha=0.5,
              position = position_jitterdodge(jitter.height = .1, jitter.width = .5))+
     scale_fill_manual(values=color_stress)+
     facet_wrap(~variable, scales = "free_y")

#save it
mapply(function(x)
  ggsave(
  plot = last_plot(),
  filename =x,
  path = "./Plots/01_alpha_beta_diversity",
  scale = 1.8,
  width = 800,
  height = 700,
  units = "px",
  dpi = 300),
x=c("alpha_phyto_ITS.svg","alpha_phyto_ITS.png"))

#Plot Herbivores
ggplot(subset(richness_long$cond_real_insects, variable %in% estimates),
       aes(x = plant_pathway_induced, y = value, fill=plant_pathway_induced))+
     geom_boxplot(aes(fill = plant_pathway_induced))+
     geom_jitter(data = subset(richness_long$cond_real_insects, variable %in% estimates),
                 mapping = aes(x = plant_pathway_induced, y = value, fill=plant_pathway_induced),
              shape = 21, size = 1.5, color = "black", alpha=0.5,
              position = position_jitterdodge(jitter.height = .1, jitter.width = .5))+
     scale_fill_manual(values=color_stress)+
     facet_wrap(~variable, scales = "free_y")

#save it
mapply(function(x)
  ggsave(
  plot = last_plot(),
  filename =x,
  path = "./Plots/01_alpha_beta_diversity",
  scale = 1.8,
  width = 800,
  height = 700,
  units = "px",
  dpi = 300),
x=c("alpha_herbivore_ITS.svg","alpha_herbivore_ITS.png"))

#Both
richness_full_long <- rbind(richness_long$cond_phytohormones, richness_long$cond_real_insects)

ggplot(subset(richness_full_long, variable %in% estimates),
       aes(x = plant_pathway_induced, y = value, fill=plant_pathway_induced))+
     geom_boxplot(aes(fill = plant_pathway_induced))+
     geom_jitter(data = subset(richness_full_long, variable %in% estimates),
                 mapping = aes(x = plant_pathway_induced, y = value, fill=plant_pathway_induced),
              shape = 21, size = 1.5, color = "black", alpha=0.5,
              position = position_jitterdodge(jitter.height = .1, jitter.width = .5))+
     scale_fill_manual(values=color_stress)+
     facet_wrap(variable~Experiment, scales = "free_y", ncol=2)

#save it
mapply(function(x)
  ggsave(
  plot = last_plot(),
  filename =x,
  path = "./Plots/01_alpha_beta_diversity",
  scale = 1.8,
  width = 1000,
  height = 1400,
  units = "px",
  dpi = 300),
x=c("alpha_both_ITS.svg","alpha_both_ITS.png"))


```

*Lower Richness in JA-induced plants in both experiments
*Simpson's D shows that phytohormone experiment was less diverse than the herbivore experiment
*PD also was lower in JA-induced plants in phyto, and with a similar trend but not so conspicuous in the herbivore-induced experiment.

Let's check if there is a correlation with library size 
```{r}

#plot it and calculate correlation coefficient
ggscatter(richness$cond_phytohormones, x = "libsize_after_filtering",
          y = estimates,
          combine=TRUE,
          ylab= "parameter",
          color="plant_pathway_induced", palette= color_stress,
          add = "reg.line",
          conf.int = TRUE,
          facet.by = NULL, scales= "free_y")+
   stat_cor(aes(color=plant_pathway_induced), method="pearson")#correlation with JA

ggscatter(richness$cond_real_insects, x = "libsize_after_filtering",
          y = estimates,
          combine=TRUE,
          ylab= "parameter",
          color="plant_pathway_induced", palette= color_stress,
          add = "reg.line",
          conf.int = TRUE,
          facet.by = NULL, scales= "free_y")+
   stat_cor(aes(color=plant_pathway_induced), method="pearson") #looks better

```
*There is a correlation between the estimates and the library size, specially for Shannon and Simpson 

*As singletons and doubletons were removed from the dataframe, Chao 1 is basically the same as just richness (Observed)

###Stats

```{r}

#make a single df with both experiments
richness_full <- rbind(richness$cond_phytohormones, richness$cond_real_insects)

#Does libsize  covary with Experiment and treatments?: Different between experiments, but no interaction
m.reads <- lm(libsize_after_filtering ~ Experiment * plant_pathway_induced,
                data=richness_full)
car::Anova(m.reads, type="III", test.statistic="F")

#check if assumptions are met
simulateResiduals(m.reads, plot=T) 
plot(m.reads)
hist(resid(m.reads)) #normalish
qqnorm(resid(m.reads)) #also

#emmeans
plot(emmeans(m.reads, pairwise ~  plant_pathway_induced | Experiment)) #overall herbivore less libsize

#clean
rm(m.reads)

```

Now I can continue analyzing each experiment independently, and focusing in Richness, Shannon and Faith's PD phylogenetic diversity

####Richness
```{r}

#change name to avoid re-writting the code
richness 

##1.Observed/Richness

obs <- lapply(richness, function(df){
        glm(df$Observed ~ plant_pathway_induced,
                data=df,
                family=Gamma(link="log"))})#changed to glm with quasipoisson distribution

obs.table <- lapply(obs, function(model){
  car::Anova(model, type="III", test.statistic="F")#using type III, as its unbalanced
})

#check if assumptions are met
lapply(obs, function(model){simulateResiduals(model, plot=T)}) #ok

#save result
factors <- c("Treatment","Residuals")
glm.p_table <- cbind (factors, obs.table$cond_phytohormones)
glm.p_table$`Pr(>F)` <- round(glm.p_table$`Pr(>F)`, 3)
table <- nice_table(glm.p_table,
              title = c("Table X. GLM for observed number of ASVs (Richness) - Phyto"),
              note = "~ plant_pathway_induced")

table <- table %>%
         flextable::colformat_num(big.mark = "", digits = 3) 

flextable::save_as_docx(table, path = "./Tables/01_alpha_beta_diversity/observed_phyto_ITS.docx")

glm.i_table <- cbind (factors, obs.table$cond_real_insects)
glm.i_table$`Pr(>F)` <- round(glm.i_table$`Pr(>F)`, 3)
table <- nice_table(glm.i_table,
              title = c("Table X. GLM for observed number of ASVs (Richness) - Insects"),
              note = "~ plant_pathway_induced")

table <- table %>%
         flextable::colformat_num(big.mark = "", digits = 3) 

flextable::save_as_docx(table, path = "./Tables/01_alpha_beta_diversity/observed_insects_ITS.docx")

#emmeans
library(multcomp)

lapply(obs, function(model){plot(emmeans(model, pairwise ~  plant_pathway_induced))}) 
obs.emmeans <- lapply(obs, function(model){
              x <- emmeans(model, pairwise ~  plant_pathway_induced)
              x <- multcomp::cld(x, adjust="sidak", Letters=letters)
              return(x)})

x <- as.data.frame(obs.emmeans$cond_phytohormones)

#make it a df  
obs.df <- lapply(obs.emmeans, function(emmeans){
  x <- as.data.frame(emmeans)
  return(x)})


```

####Shannon
```{r}

descdist(richness$cond_phytohormones$Shannon, discrete = FALSE)

##2.Shannon
shann <- lapply(richness, function(df){
          lm(log(df$Shannon) ~ plant_pathway_induced,
                data=df)})

shann.table <- lapply(shann, function(model){
  car::Anova(model, type="III", test.statistic="F")#using type III, as its unbalanced
})

shann.table #both not significant

#check if assumptions are met
lapply(shann, function(model){simulateResiduals(model, plot=T)}) #ok


#save result
factors <- c("Intercept","Treatment","Residuals")
lm.p_table <- cbind (factors, shann.table$cond_phytohormones)
lm.p_table$`Pr(>F)` <- round(lm.p_table$`Pr(>F)`, 3)
table <- nice_table(lm.p_table,
              title = c("Table X. LM for Shannon - Phyto"),
              note = "~ plant_pathway_induced")

table <- table %>%
         flextable::colformat_num(big.mark = "", digits = 3) 

flextable::save_as_docx(table, path = "./Tables/01_alpha_beta_diversity/shannon_phyto_ITS.docx")

lm.i_table <- cbind (factors, shann.table$cond_real_insects)
lm.i_table$`Pr(>F)` <- round(lm.i_table$`Pr(>F)`, 3)
table <- nice_table(lm.i_table,
              title = c("Table X. LM for Shannon - Insects"),
              note = "~ plant_pathway_induced")

table <- table %>%
         flextable::colformat_num(big.mark = "", digits = 3) 

flextable::save_as_docx(table, path = "./Tables/01_alpha_beta_diversity/shannon_insects_ITS.docx")

```

No difference, same shannon index value (and Simpson, very similar communities)

####Faith's PD
```{r}

##3. Phylogenetic Diversity
pd <- lapply(richness, function(df){
        glm(log(df$PD) ~ plant_pathway_induced,
                data=df,
            family=Gamma(link="log"))})

summary(pd$cond_phytohormones)

pd.table <- lapply(pd, function(model){
  car::Anova(model, type="III", test.statistic="LR")#using type III, as its unbalanced
})

pd.table #both signigicant

#check if assumptions are met
lapply(pd, function(model){simulateResiduals(model, plot=T)}) #ok

#save result
factors <- c("Treatment")
lm.p_table <- cbind (factors, pd.table$cond_phytohormones)
lm.p_table$`Pr(>Chisq)` <- round(lm.p_table$`Pr(>Chisq)`, 3)
table <- nice_table(lm.p_table,
              title = c("Table X. GLM for Faith's PD - Phyto"),
              note = "~ plant_pathway_induced, GLM with log link")

table <- table %>%
         flextable::colformat_num(big.mark = "", digits = 3) 

flextable::save_as_docx(table, path = "./Tables/01_alpha_beta_diversity/PD_phyto_ITS.docx")

lm.i_table <- cbind (factors, pd.table$cond_real_insects)
lm.i_table$`Pr(>Chisq)` <- round(lm.i_table$`Pr(>Chisq)`, 3)
table <- nice_table(lm.i_table,
              title = c("Table X. GLM for Faith's PD - Insects"),
              note = "~ plant_pathway_induced")

table <- table %>%
         flextable::colformat_num(big.mark = "", digits = 3) 

flextable::save_as_docx(table, path = "./Tables/01_alpha_beta_diversity/PD_insects_ITS.docx")

#emmeans
lapply(pd, function(model){plot(emmeans(model, pairwise ~  plant_pathway_induced))}) 
pd.emmeans <- lapply(pd, function(model){
              x <- emmeans(model, pairwise ~  plant_pathway_induced)
              x <- multcomp::cld(x, adjust="none", Letters=letters)
              return(x)})

x <- as.data.frame(obs.emmeans$cond_phytohormones)

#make it a df  
obs.df <- lapply(obs.emmeans, function(emmeans){
  x <- as.data.frame(emmeans)
  return(x)})


```






####Richness
```{r}






#run the model 
lm <- lm(df$Observed ~ plant_pathway_induced + libsize_after_filtering,
                data=df)

#look at the results 
#Option 1
car::Anova(model, type="III", test.statistic="F")

#Option 2
summary(lm)

#check if assumptions are met
simulateResiduals(model, plot=T) #using DHARMa and fitdistrplus package 
hist(resid(model)) #normalish
qqnorm(resid(model)) #also

#see differences between treatments 
plot(emmeans(lm, pairwise ~  plant_pathway_induced)) #emmean package 
obs.emmeans <- emmeans(lm, pairwise ~  plant_pathway_induced)

#change it to df 
df_lm <- as.data.frame(lm) #not sure, you have to check if this is the way
df_emmeans <- obs.emmeans$contrasts %>% summary(infer=TRUE) %>% as.data.frame()

#save results with rempsyc package
table_lm <- nice_table(df_lm,
              title = c("Table X. Results of Linear Model testing alpha diversity metrics for ITS communities"))
              
flextable::save_as_docx(table_lm, path =  "./Tables/01_alpha_beta_diversity/both_alpha_lm_ITS.docx") #saves the table in word

#repeat for emmeans results 




##1.Observed/Richness
obs <- lapply(richness, function(df){
        lm(df$Observed ~ plant_pathway_induced + libsize_after_filtering,
                data=df)})

obs.table <- lapply(obs, function(model){
  car::Anova(model, type="III", test.statistic="F")#using type III, as its unbalanced
})

#check if assumptions are met
lapply(obs, function(model){simulateResiduals(model, plot=T)}) #ok
lapply(obs, function(model){hist(resid(model))}) #normalish
lapply(obs, function(model){qqnorm(resid(model))}) #also

#emmeans: Post-hoc test
lapply(obs, function(model){plot(emmeans(model, pairwise ~  plant_pathway_induced))}) 
obs.emmeans <- lapply(obs, function(model){
              emmeans(model, pairwise ~  plant_pathway_induced)})

#make it a df  
obs.df <- lapply(obs.emmeans, function(emmeans){
  emmeans$contrasts %>% summary(infer=TRUE) %>% as.data.frame()})

```

####Shannon
```{r}

##2.Shannon
shann <- lapply(richness, function(df){
        lm(df$Shannon ~ plant_pathway_induced + libsize_after_filtering,
                data=df)})

shann.table <- lapply(shann, function(model){
  car::Anova(model, type="III", test.statistic="F")#using type III, as its unbalanced
})

shann.table #both signigicant

#check if assumptions are met
lapply(shann, function(model){simulateResiduals(model, plot=T)}) #ok
lapply(shann, function(model){hist(resid(model))}) #normalish
lapply(shann, function(model){qqnorm(resid(model))}) #also

#emmeans: Post-hoc test
lapply(shann, function(model){plot(emmeans(model, pairwise ~  plant_pathway_induced))}) 
shann.emmeans <- lapply(shann, function(model){
              emmeans(model, pairwise ~  plant_pathway_induced)})

#df  
shann.df <- lapply(shann.emmeans, function(emmeans){
  emmeans$contrasts %>% summary(infer=TRUE) %>% as.data.frame()})

```

Difference between control and SA only in Insects experiment.

####Faith's PD
```{r}

##3. Phylogenetic Diversity
pd <- lapply(richness, function(df){
        lm(df$PD ~ plant_pathway_induced + libsize_after_filtering,
                data=df)})

pd.table <- lapply(pd, function(model){
  car::Anova(model, type="III", test.statistic="F")#using type III, as its unbalanced
})

pd.table #both signigicant

#check if assumptions are met
lapply(pd, function(model){simulateResiduals(model, plot=T)}) # not ok for phyto
lapply(pd, function(model){hist(resid(model))}) #normalish
lapply(pd, function(model){qqnorm(resid(model))}) #also

#emmeans: Post-hoc test
lapply(pd, function(model){plot(emmeans(model, pairwise ~  plant_pathway_induced))}) 
pd.emmeans <- lapply(pd, function(model){
              emmeans(model, pairwise ~  plant_pathway_induced)})

#df  
pd.df <- lapply(pd.emmeans, function(emmeans){
  emmeans$contrasts %>% summary(infer=TRUE) %>% as.data.frame()})

```

No difference between treatments and controls 

###Results tables

Format p value: 
https://scriptsandstatistics.wordpress.com/2019/09/22/formating-p-values-a-curated-list-of-r-functions/
```{r}

#1. Get linear model results
#models
obs.table
shann.table
pd.table

#make a function to extract values into a df to not repeat the code 
make_lm_table_2_experiments <-function(model.table, name_index){
  
 #get name of index
 index <- deparse(substitute(name_index))
  
#split in two
 phyto <-  model.table$cond_phytohormones
 herbivore <- model.table$cond_real_insects
 
#rename each 
 p <- phyto %>% rename(P.value=`Pr(>F)`)%>% 
           rename_with(~paste0("p_", .x))
 p$p_P.value <- scales::pvalue(p$p_P.value)#round p value

 h <- herbivore %>% rename(P.value=`Pr(>F)`)%>% 
           rename_with(~paste0("h_", .x))
 h$h_P.value <- scales::pvalue(h$h_P.value)#round p value
 
 both <- cbind(p,h)
 both <- both %>% mutate(Index=index)%>%
                  rownames_to_column(var="covariates")%>%
                  relocate(Index)
 return(both)
}

#create dataframes
obs_both <-make_lm_table_2_experiments(obs.table, Richness)
shann_both <-make_lm_table_2_experiments(shann.table, Shannon)
pd_both <-make_lm_table_2_experiments(pd.table, `Faith's PD`)

#Finally, merge them
alpha_df <- rbind(obs_both, shann_both, pd_both)

#save it
table <- nice_table(alpha_df,
              title = c("Table X. Results of Linear Model testing alpha diversity metrics for ITS communities"))
              
flextable::save_as_docx(table, path = "./Tables/01_alpha_beta_diversity/both_alpha_lm_ITS.docx")

#2. Get post-hoc results
obs.df
shann.df
pd.df

#make a function to not repeat the code 
make_emmeans_table_2_experiments <-function(emmeans.df, name_index){
  
 #get name of index
 index <- deparse(substitute(name_index))
  
#split in two
 phyto <-  emmeans.df$cond_phytohormones
 herbivore <- emmeans.df$cond_real_insects
 
#rename each 
 p <- phyto %>% column_to_rownames(var = "contrast") %>% 
   rename_with(~paste0("p_", .x))
 p$p_p.value <- scales::pvalue(p$p_p.value)#round p value

 h <- herbivore %>% column_to_rownames(var = "contrast") %>% 
   rename_with(~paste0("h_", .x))
 h$h_p.value <- scales::pvalue(h$h_p.value)#round p value
 
 both <- cbind(p,h)
 both <- both %>% mutate(Index=index)%>%
                  rownames_to_column(var="contrast")%>%
                  relocate(Index, contrast)
 return(both)
}

obs_emm_both <- make_emmeans_table_2_experiments(obs.df, Richness)
shann_emm_both <- make_emmeans_table_2_experiments(shann.df, Shannon)
pd_emm_both <- make_emmeans_table_2_experiments(pd.df, `Faith's PD`)

#Finally, merge them
alpha_posthoc_df <- rbind(obs_emm_both, shann_emm_both, pd_emm_both)

#save it
table <- nice_table(alpha_posthoc_df,
              title = c("Table X. Results of Emmeans testing alpha diversity metrics for ITS communities"))
              
flextable::save_as_docx(table, path = "./Tables/01_alpha_beta_diversity/both_alpha_ph_ITS.docx")

#3. Finally, save df with alpha diversity metrics results
richness_full
write.csv(richness_full, file="./Data_frames/01_alpha_beta_diversity/alpha_metrics_ITS.csv", row.names = TRUE)

#ready!

#clean
rm(list=ls())

```

#16S

Continue with 16S, go back to line 12 to load packages and plot settings again

##Loading data
```{r}
#16S
load(file = "./amplicon_data/Phyloseq_objects/ps_16S.RData")
ps_16S

#let's quickly check how many replicates we have left per experiment
#and how is the library size amongst them
df_summ <- ps_16S %>% 
    sample_data() %>% 
    group_by(Experiment, Stress)%>%
    summarise(Replicates = n(),
              library_size = mean(libsize_after_filtering),
              stdev_libsize =sd(libsize_after_filtering)) %>% 
    as.data.frame()

df_summ%>%
    ggplot(., aes(x=Stress,y=library_size, label=Replicates))+
    geom_bar(stat = "identity")+
    geom_text(size=8)+
    facet_wrap(~Experiment, scales = "free_x")
#around same number of reps but Insect herbivory exp has overall less reads 

#save this table
table <- nice_table(df_summ,
              title = c("Table X. Summary for 16S"))
flextable::save_as_docx(table, path = "./Tables/01_alpha_beta_diversity/summary_16S.docx")

#get sample data df 
sample_data <- as.data.frame(sample_data(ps_16S))


#Load 16S split by experiment
load(file = "./amplicon_data/Phyloseq_objects/ps_16S_split.RData")
p16S_exp <- ps_16S_split
p16S_exp <- p16S_exp[-3] #let's remove EPG for now

#check sample sums
lapply(p16S_exp, function(ps){sort(sample_sums(ps))}) 

#Melissa's median transformation
#total = median(sample_sums(physeq_f_unnorm))
#standf = function(x, t=total) round(t * (x / sum(x)))
#physeq_msd = transform_sample_counts(physeq_f_unnorm, standf)

#let's do this analysis at family level
#p16S_family <- lapply(p16S_exp, function(ps){
#  ps <- tax_glom(ps, taxrank="Family")
#  return(ps)
#})


#transform counts with median transformation (recommended by Melissa)
p16S_median <- lapply(p16S_exp, function(ps){
  total <- median(sample_sums(ps))
  standf <- function(x, t=total) round(t * (x / sum(x)))
  physeq_msd = transform_sample_counts(ps, standf)
  print((sample_sums(physeq_msd)))
  return(physeq_msd)
})

#rarefy to even depth
#p16S_exp_rar <-lapply(p16S_exp, function(ps){
#  rarefy_even_depth(ps, sample.size = min(sample_sums(ps)),
#  rngseed = 123, replace = TRUE, trimOTUs = TRUE, verbose = TRUE)})

#lapply(p16S_exp_rar, function(ps){ sample_sums(ps)}) 
#49,633 reads for Phyto
#20,706 reads for Insects

```

More or less same replicates 

##Alpha diversity
```{r}

#plot it
#p <- lapply(p16S_median, function(ps){plot_richness(ps, x= "plant_pathway_induced", scales="free_y")})

#lapply(p, function(plot){
#  plot + geom_boxplot(data=plot$data, aes(x= Experiment, y=value, color=NULL), alpha=0.1)+
#    scale_color_manual(values=color_experiment)})

#Observed(Richness):Total number of ASVs
#Chao1:Same as Observed but aims to calculate real number of ASVs using single- and doubletons, cannot be used
#Shannon(diversity index):Lower value, lower diversity
#Simpson(similarity index):Takes abundance (eveness) into account. Lower value, higher diversity


#get Observed, Shannon and Simpson indexes
richness_16S <- lapply(p16S_median, function(ps){
  estimate_richness(ps,
            measures = c("Observed", "Shannon", "Simpson"))
})

#fix df 
richness_16S <- lapply(richness_16S, function(df){
  rownames(df) <- sub("X", "", rownames(df)) #removes X from rownames
  df$sample_name <- rownames(df) 
  return(df)
})

#Get Faith's phylogenetic diversity index
pd <- lapply(p16S_median, function(ps){
  pd(t(as.data.frame(ps@otu_table)), ps@phy_tree,include.root=T)
  })

#merge with metadata of ps object
richness_16S$cond_phytohormones <- cbind(as.data.frame(sample_data(p16S_median$cond_phytohormones)), richness_16S$cond_phytohormones, pd$cond_phytohormones)

richness_16S$cond_real_insects<- cbind(as.data.frame(sample_data(p16S_median$cond_real_insects)), richness_16S$cond_real_insects, pd$cond_real_insects)

#select the columns needed
richness_16S <- lapply(richness_16S, function(df){
  df <- df %>% dplyr::select(c('Experiment',
                         'plant_pathway_induced',
                         'Amplicon_concentation_ngul',
                         'total_reads',
                         'libsize_after_filtering',
                         'shoot_biomass',"Observed",
                         "Shannon", "Simpson", "PD", "sample_name"))
})

#make another set with long format for plotting
richness_long_16S <- lapply(richness_16S, function(df){
  df_long <- reshape2::melt(df)})

#clean
rm(p, pd)

```

For now, just work with Observed(Richness), Shannon and PD

###Plot

Plot it
```{r}

estimates <- c("Observed","Shannon","PD")

#add info for y axis

limits <- data.frame(
  variable = c("Observed","Shannon","PD"),
  y_min = c(0, 4, 50), # Set your desired minimum y-axis limits
  y_max = c(3000, 8, 180) # Set your desired maximum y-axis limits
)

#merge
richness_long_16S_estimates <- lapply(richness_long_16S, function(df){
  df <- filter(df, variable %in% estimates)
  df <- df %>% left_join(limits, by="variable")
  df$variable <- factor(df$variable, levels=c("Observed","Shannon","PD"))
  return(df)
  })


#Plot phytohormones
p.p <- ggplot(richness_long_16S_estimates$cond_phytohormones,
       aes(x = plant_pathway_induced, y = value, fill=plant_pathway_induced))+
     geom_boxplot(aes(fill = plant_pathway_induced))+
     geom_jitter(data = richness_long_16S_estimates$cond_phytohormones,
                 mapping = aes(x = plant_pathway_induced, y = value, fill=plant_pathway_induced),
              shape = 21, size = 1.5, color = "black", alpha=0.5,
              position = position_jitterdodge(jitter.height = .1, jitter.width = .5))+
     scale_fill_manual(values=color_stress)+
  geom_line() +
  geom_blank(aes(y = y_min)) + # Force the minimum y limit
  geom_blank(aes(y = y_max)) + # Force the maximum y limit
  facet_wrap(~ variable, scales = "free_y") +
  theme(legend.position = "none")+
    axis_looks

p.i <- ggplot(richness_long_16S_estimates$cond_real_insects,
       aes(x = plant_pathway_induced, y = value, fill=plant_pathway_induced))+
     geom_boxplot(aes(fill = plant_pathway_induced))+
     geom_jitter(data = richness_long_16S_estimates$cond_real_insects,
                 mapping = aes(x = plant_pathway_induced, y = value, fill=plant_pathway_induced),
              shape = 21, size = 1.5, color = "black", alpha=0.5,
              position = position_jitterdodge(jitter.height = .1, jitter.width = .5))+
     scale_fill_manual(values=color_stress)+
  geom_line() +
  geom_blank(aes(y = y_min)) + # Force the minimum y limit
  geom_blank(aes(y = y_max)) + # Force the maximum y limit
  facet_wrap(~ variable, scales = "free_y") +
  theme(legend.position = "none")+
    axis_looks


library(patchwork)

p.p / p.i

#save it
mapply(function(x)
  ggsave(
  plot = last_plot(),
  filename =x,
  path = "./Plots/01_alpha_beta_diversity",
  scale = 2.0,
  width = 1000,
  height = 800,
  units = "px",
  dpi = 300),
x=c("alpha_both_16S.pdf","alpha_both_16S.png"))

```

*No clear patterns

Let's check if there is a correlation with library size 
```{r}

#plot it and calculate correlation coefficient
ggscatter(richness_16S$cond_phytohormones, x = "libsize_after_filtering",
          y = estimates,
          combine=TRUE,
          ylab= "parameter",
          color="plant_pathway_induced", palette= color_stress,
          add = "reg.line",
          conf.int = TRUE,
          facet.by = NULL, scales= "free_y")+
   stat_cor(aes(color=plant_pathway_induced), method="pearson")#correlation with Observed and Shannon

ggscatter(richness_16S$cond_real_insects, x = "libsize_after_filtering",
          y = estimates,
          combine=TRUE,
          ylab= "parameter",
          color="plant_pathway_induced", palette= color_stress,
          add = "reg.line",
          conf.int = TRUE,
          facet.by = NULL, scales= "free_y")+
   stat_cor(aes(color=plant_pathway_induced), method="pearson") #Correlates with the 3 indexes

```
*There is a correlation between the estimates and the library size, specially for Shannon and Simpson 


###Stats
```{r}

#make a single df with both experiments
richness_full_16S <- rbind(richness_16S$cond_phytohormones, richness_16S$cond_real_insects)

#Does libsize  covary with Experiment and treatments?: Different between experiments, but no interaction
m.reads <- glm(libsize_after_filtering ~ Experiment * plant_pathway_induced,
                data=richness_full_16S,
               family=Gamma("inverse")) #skewed to the left

car::Anova(m.reads, type="III", test.statistic="F")

#check if assumptions are met
simulateResiduals(m.reads, plot=T) #ok
plot(m.reads)
hist(resid(m.reads)) #better
qqnorm(resid(m.reads)) #also

#emmeans
plot(emmeans(m.reads, pairwise ~  plant_pathway_induced | Experiment))#nothing different

#clean
rm(m.reads)

```

Now I can continue analyzing each experiment independently, and focusing in Richness, Shannon and Faith's PD phylogenetic diversity

####Richness
```{r}

#change name to avoid re-writting the code
richness <- richness_16S

##1.Observed/Richness

obs <- lapply(richness, function(df){
        glm(df$Observed ~ plant_pathway_induced,
                data=df,
                family=Gamma(link="log"))})#changed to glm with quasipoisson distribution

obs.table <- lapply(obs, function(model){
  car::Anova(model, type="III", test.statistic="F")#using type III, as its unbalanced
})

#check if assumptions are met
lapply(obs, function(model){simulateResiduals(model, plot=T)}) #ok

#save result
factors <- c("Treatment","Residuals")
glm.p_table <- cbind (factors, obs.table$cond_phytohormones)
glm.p_table$`Pr(>F)` <- round(glm.p_table$`Pr(>F)`, 3)
table <- nice_table(glm.p_table,
              title = c("Table X. GLM for observed number of ASVs (Richness) - Phyto"),
              note = "~ plant_pathway_induced")

table <- table %>%
         flextable::colformat_num(big.mark = "", digits = 3) 

flextable::save_as_docx(table, path = "./Tables/01_alpha_beta_diversity/observed_phyto.docx")

glm.i_table <- cbind (factors, obs.table$cond_real_insects)
glm.i_table$`Pr(>F)` <- round(glm.i_table$`Pr(>F)`, 3)
table <- nice_table(glm.i_table,
              title = c("Table X. GLM for observed number of ASVs (Richness) - Insects"),
              note = "~ plant_pathway_induced")

table <- table %>%
         flextable::colformat_num(big.mark = "", digits = 3) 

flextable::save_as_docx(table, path = "./Tables/01_alpha_beta_diversity/observed_insects.docx")

```

####Shannon
```{r}

descdist(richness$cond_phytohormones$Simpson, discrete = FALSE)

##2.Shannon
shann <- lapply(richness, function(df){
        lm(log(df$Shannon) ~ plant_pathway_induced,
                data=df)})

shann.table <- lapply(shann, function(model){
  car::Anova(model, type="III", test.statistic="F")#using type III, as its unbalanced
})

shann.table #both not significant

#check if assumptions are met
lapply(shann, function(model){simulateResiduals(model, plot=T)}) #ok


#save result
factors <- c("Intercept","Treatment","Residuals")
lm.p_table <- cbind (factors, shann.table$cond_phytohormones)
lm.p_table$`Pr(>F)` <- round(lm.p_table$`Pr(>F)`, 3)
table <- nice_table(lm.p_table,
              title = c("Table X. LM for Shannon - Phyto"),
              note = "~ plant_pathway_induced")

table <- table %>%
         flextable::colformat_num(big.mark = "", digits = 3) 

flextable::save_as_docx(table, path = "./Tables/01_alpha_beta_diversity/shannon_phyto.docx")

lm.i_table <- cbind (factors, shann.table$cond_real_insects)
lm.i_table$`Pr(>F)` <- round(lm.i_table$`Pr(>F)`, 3)
table <- nice_table(lm.i_table,
              title = c("Table X. LM for Shannon - Insects"),
              note = "~ plant_pathway_induced")

table <- table %>%
         flextable::colformat_num(big.mark = "", digits = 3) 

flextable::save_as_docx(table, path = "./Tables/01_alpha_beta_diversity/shannon_insects.docx")

```

No difference, same shannon index value (and Simpson, very similar communities)

####Faith's PD
```{r}

##3. Phylogenetic Diversity
pd <- lapply(richness, function(df){
        glm(log(df$PD) ~ plant_pathway_induced,
                data=df,
            family=Gamma(link="log"))})

pd.table <- lapply(pd, function(model){
  car::Anova(model, type="III", test.statistic="F")#using type III, as its unbalanced
})

pd.table #both signigicant

#check if assumptions are met
lapply(pd, function(model){simulateResiduals(model, plot=T)}) #ok


#save result
factors <- c("Treatment","Residuals")
lm.p_table <- cbind (factors, pd.table$cond_phytohormones)
lm.p_table$`Pr(>F)` <- round(lm.p_table$`Pr(>F)`, 3)
table <- nice_table(lm.p_table,
              title = c("Table X. GLM for Faith's PD - Phyto"),
              note = "~ plant_pathway_induced")

table <- table %>%
         flextable::colformat_num(big.mark = "", digits = 3) 

flextable::save_as_docx(table, path = "./Tables/01_alpha_beta_diversity/PD_phyto.docx")

lm.i_table <- cbind (factors, pd.table$cond_real_insects)
lm.i_table$`Pr(>F)` <- round(lm.i_table$`Pr(>F)`, 3)
table <- nice_table(lm.i_table,
              title = c("Table X. GLM for Faith's PD - Insects"),
              note = "~ plant_pathway_induced")

table <- table %>%
         flextable::colformat_num(big.mark = "", digits = 3) 

flextable::save_as_docx(table, path = "./Tables/01_alpha_beta_diversity/PD_insects.docx")


#emmeans
lapply(pd, function(model){plot(emmeans(model, pairwise ~  plant_pathway_induced))}) 
pd.emmeans <- lapply(pd, function(model){
              x <- emmeans(model, pairwise ~  plant_pathway_induced)
              x <- multcomp::cld(x, adjust="sidak", Letters=letters)
              return(x)})

#make it a df  
pd.df <- lapply(pd.emmeans, function(emmeans){
  x <- as.data.frame(emmeans)
  return(x)})

```

No difference between treatments and controls in any of the experiments 

###Results tables

Format p value: 
https://scriptsandstatistics.wordpress.com/2019/09/22/formating-p-values-a-curated-list-of-r-functions/
```{r}

#1. Get linear model results
#models
obs.table
shann.table
pd.table

#make a function to extract values into a df to not repeat the code 
make_lm_table_2_experiments <-function(model.table, name_index){
  
 #get name of index
 index <- deparse(substitute(name_index))
  
#split in two
 phyto <-  model.table$cond_phytohormones
 herbivore <- model.table$cond_real_insects
 
#rename each 
 p <- phyto %>% rename(P.value=`Pr(>F)`)%>% 
           rename_with(~paste0("p_", .x))
 p$p_P.value <- scales::pvalue(p$p_P.value)#round p value

 h <- herbivore %>% rename(P.value=`Pr(>F)`)%>% 
           rename_with(~paste0("h_", .x))
 h$h_P.value <- scales::pvalue(h$h_P.value)#round p value
 
 both <- cbind(p,h)
 both <- both %>% mutate(Index=index)%>%
                  rownames_to_column(var="covariates")%>%
                  relocate(Index)
 return(both)
}

#create dataframes
obs_both <-make_lm_table_2_experiments(obs.table, Richness)
shann_both <-make_lm_table_2_experiments(shann.table, Shannon)
pd_both <-make_lm_table_2_experiments(pd.table, `Faith's PD`)

#Finally, merge them
colnames(shann_both) <- colnames(obs_both)
alpha_df_16S <- rbind(obs_both, shann_both, pd_both)

#save it
table <- nice_table(alpha_df_16S,
              title = c("Table X. Results of Linear Model testing alpha diversity metrics for 16S communities"))
              
flextable::save_as_docx(table, path = "./Tables/01_alpha_beta_diversity/both_alpha_lm_16S.docx")

#2. Get post-hoc results (not necessary as nothing was significant)

#3. Finally, save df with alpha diversity metrics results
richness_full_16S
write.csv(richness_full_16S, file="./Data_frames/01_alpha_beta_diversity/alpha_metrics_16S.csv", row.names = TRUE)

#save environment
save.image(file='./R_environments/01_alpha_beta_diversity/alpha_16S.rdata')

#clean
rm(list=ls())

```



