---
title: "Differential abundance"
author: "Marcela Aragon"
date: "2024-08-02"
output: html_document
---

Continue with differential abundance analysis, one per experiment as you've been doing. Follow the same code that you already have, don't get creative!!

Use the ps_ITS object in Results/pre-procesing ITS and that one split it by experiment, it's not a big deal at this point that the filtering was not done by experiment as probably the same samples are going to be missing. Later with time, come back to the ps_ITS, for now it is forbidden!! 

#Load libraries

```{r}

#Load libraries

#microbiome data
library(phyloseq)
library(microbiome)
library(metagMisc)
library(metagenomeSeq)#CSS normalization
library(vegan)
library(ape)

#data handling
library(dplyr)
library(tidyr)
library(tibble)
library(readxl)
library(purrr) #melts df of lists into a single one
library(varhandle)

#stats
library(stats) #stats
library(rempsyc) #tables
library(scales)#rounds p value
library(EcolUtils) #pairwise Adonis

#ploting
library(ggpubr)
library(ggplot2)
library(RColorBrewer) #nice colors for plots
library(UpSetR)
library(ComplexHeatmap)
library(scran)
library(cowplot)
library(VennDiagram)
library(pscl)
library(MASS)
library(phytools)
library(VennDiagram)
library(itol.toolkit)

#Differential Abundance
library(ANCOMBC) #ANCOM

```

## Settings for plots
```{r}

#Set plots theme
ggplot2::theme_set(theme_bw())

#Set colors
color_stress <- c("#BFBFBF", "#E69F00", "#4472C4")
color_experiment <- c("#CC3399", "#70AD47")

color_ctrl_vs_JA <- c("#BFBFBF", "#E69F00")
color_ctrl_vs_SA <- c("#BFBFBF",  "#4472C4")
color_JA <- "#E69F00"
color_SA <- "#4472C4"
color_experiment <- c("#CC3399", "#70AD47")
fonts <- windowsFonts()

```

#16S

```{r}
#Load data

#16S
load(file = "./amplicon_data/Phyloseq_objects/ps_16S_split.RData")
ps_16S_split

```


##Pre-filtering

```{r}
#First, make a pre-filtering of those taxa that are low in prevalence (<20%)

ps_16S_filt <- lapply(ps_16S_split, function(ps){
  metagMisc::phyloseq_filter_prevalence(ps, prev.trh = 0.20, #at least 20% of samples
                                          abund.trh = NULL) #no filtering by abundance
})

#check
ps_16S_filt #less than 2,000 taxa

```

##ANCOM
```{r}

set.seed(123)


ancom <- lapply(ps_16S_split, function(ps){
  set.seed(123)
  ANCOMBC::ancombc(phyloseq = ps, 
                      formula = "plant_pathway_induced", 
                      p_adj_method = "fdr", #same as "BH" Benjamini & Hochberg
                      prv_cut = 0.20,
                      group = "plant_pathway_induced", 
                      struc_zero = TRUE, #with structural zero's
                      neg_lb = FALSE, #classifies structural zero's if it's zero and not its asymptotic lower bound 
                      tol = 1e-5,               
                      max_iter = 100, 
                      conserve = TRUE, #conservative variance of the test statistics 
                      alpha = 0.05,
                      verbose=TRUE,
                      global=FALSE)})

save(ancom, file = "./Results/03_differential_abundance/ancom_16S.Rdata")

```

Get results 
```{r}

head(ancom$cond_phytohormones$feature_table)

#check
lapply(ancom, function(ancom_result){head(ancom_result$res$lfc)})#LFC
lapply(ancom, function(ancom_result){head(ancom_result$res$se)})#standard error
lapply(ancom, function(ancom_result){head(ancom_result$res$W)}) #test statistic LFC/se

#get sample fractions
samp_frac <- lapply(ancom, function(ancom_result){ancom_result$samp_frac})#ancom$samp_frac

# Get log transformed counts and add pseudo-count (1) to feature table to avoid taking the log of 0
log_obs_abn <-lapply(ancom, function(ancom_result){log(ancom_result$feature_table +1)})

# Adjust the log observed abundances with the sample fraction
log_corr_abn <-mapply(function(x,y){
                     t(t(x) - y)},
                     x=log_obs_abn,
                     y=samp_frac)
  
#Results
log_corr_abn #bias-corrected log abundances
lfc <- lapply(ancom, function(ancom_result){ancom_result$res$lfc}) #log-fold change according to control 
se_lfc <- lapply(ancom, function(ancom_result){ancom_result$res$se}) #standard error of log-fold change

#get significant ASVs   
JA_asvs <- lapply(ancom, function(ancom_result){
           asvs <- ancom_result$res$diff_abn %>%
            dplyr::filter(plant_pathway_inducedJA == TRUE) #26,12,8
           asvs$ASV <- asvs$taxon
           return(asvs)})

SA_asvs <- lapply(ancom, function(ancom_result){
           asvs <- ancom_result$res$diff_abn %>%
            dplyr::filter(plant_pathway_inducedSA == TRUE) #9,4,10
           asvs$ASV <- asvs$taxon
           return(asvs)})

#Now I need to get the unique names 
sig_asvs <- mapply(function(x,y){
  c(x$ASV, y$ASV)},
  x=JA_asvs,
  y=SA_asvs)

#make it into a dataframe format
sig_asvs <- lapply(sig_asvs, function(x){
   unique <- unique(x)
   unique <- as.data.frame(unique)
   colnames(unique) <- "ASV"
   return(unique)})  

#With this I can slice the different data frames to plot them

#1.bias-corrected abundances
log_corr_abn_sig <- mapply(function(x,y){x[y$ASV,]},
  x=log_corr_abn,
  y=sig_asvs) #works because its a matrix 

#2.log-fold change vs controls
lfc_sig <- mapply(function(x,y){x[x$taxon %in% y$ASV,]},
                  x=lfc,
                  y=sig_asvs,
                  SIMPLIFY = FALSE) #to keep it as a list and not a single thing

#3.Differential abundance according to stress 
dif_abun_stress_sig <- mapply(function(x,y){x$res$diff_abn[x$res$diff_abn$taxon %in% y$ASV,]},
                           x=ancom,
                           y=sig_asvs,
                           SIMPLIFY = FALSE) 

#place ASV as rowname to avoid mistakes later
dif_abun_stress_sig <- lapply(dif_abun_stress_sig, function(df){
  rownames(df) <- df$taxon
  df <- df[,3:4]
  return(df)
})

```

##Heatmap LFC by treatment

To make my life easier I'll plot one per experiment and just run the same code 

###Phyto
```{r}

#I need:
lfc_sig$cond_phytohormones #values of lfc 
dif_abun_stress_sig$cond_phytohormones #info if DA in JA and/or SA
ps_16S_split$cond_phytohormones #phyloseq

#change name to run the same code 
lfc <- lfc_sig$cond_phytohormones
dif_abun_stress <- dif_abun_stress_sig$cond_phytohormones

#let's plot only the LFC 

#Now, let's see what happens when the LFC results from ANCOMBC are plotted in a heatmap
rownames(lfc) <- lfc$taxon
lfc <- lfc[,3:4]

#make a matrix from the lfc differences
matrix.log.phyto <- as.matrix(lfc) 

#make new ps with only the DA asvs
ps_DA <- phyloseq::prune_taxa(sig_asvs$cond_phytohormones$ASV, ps_16S_split$cond_phytohormones)

#get species name if available, if not replace with name of ASV
df_row <- as.data.frame((tax_table(ps_DA))[, c("Genus")])
df_row <- df_row %>%
           mutate(asv_name=rownames(.)) #new column with rownames
df_row$Genus <- ifelse(is.na(df_row$Genus), df_row$asv_name, df_row$Genus)
df_row <- df_row %>%
           dplyr::select(Genus)

#order data so it matches
dif_abun_stress_ordered <- dif_abun_stress[order(match(rownames(dif_abun_stress),
                                                     rownames(tax_table(ps_DA)))), , drop=FALSE] 

matrix.log.phyto_ordered <- matrix.log.phyto[order(match(rownames(matrix.log.phyto),
                                                     rownames(tax_table(ps_DA)))), , drop=FALSE] 

#define annotation for rows
annotation_row = data.frame(
    Phylum = as.factor(tax_table(ps_DA)[, "Phylum"]), #to add phylum in the heatmap as well
    Contrast = dif_abun_stress_ordered) 
colnames(annotation_row)[2:3] <- c("JAvsCtrl", "SAvsCtrl")

#change it to factor so it can be colored
annotation_row$JAvsCtrl <- as.factor(annotation_row$JAvsCtrl)
annotation_row$SAvsCtrl <- as.factor(annotation_row$SAvsCtrl)

#check if rownames are in same order 
rownames(tax_table(ps_DA)) #annotation row
rownames(df_row) #yes

#check if names are the same 
rownames(matrix.log.phyto_ordered)
rownames(annotation_row) #yes

#make labels for rows 
df_row
species_ordered <- df_row[order(match(rownames(df_row), rownames(matrix.log.phyto_ordered))), , drop = FALSE]

# select colors for factors
#make colors for phylum with custom function
colourCount <- length(levels(annotation_row$Phylum))
getPalette <- colorRampPalette(brewer.pal(colourCount, "Set3"))

#apply to phylum
phylum_col = getPalette(colourCount)
names(phylum_col) = levels(annotation_row$Phylum)
names(color_experiment) = c("cond_phytohormones", "real_insects")

#put all colors together

color_stress <- c("#BFBFBF", "#E69F00", "#4472C4")
color_experiment <- c("#CC3399", "#70AD47")

annotation_colors = list(
     Phylum = phylum_col,
     JAvsCtrl = c("TRUE" = "#E69F00", "FALSE" = "white"),
     SAvsCtrl = c("TRUE" = "#4472C4", "FALSE" = "white"),
     Experiment = color_experiment)

#make colors for matrix
#Select colors for the ends of the values and the middle 
function_color_matrix <- circlize::colorRamp2(c(-2.5, 0, 2.5),
                          c("#0088c9",  "white", "#c85200"))
                         # c("#0000FF", "white", "#FF0000")) #palette blue/red
                         # c("#018571", "white", "#A6611A")) #palette teal/brown

function_color_matrix(seq(-3,3)) #gives you 9 colors with white in the middle 

# now plot the pheatmap
lfc_heatmap <- ComplexHeatmap::pheatmap(matrix.log.phyto_ordered,
                         color= function_color_matrix,
                         name="LFC",
                         scale= "none", 
                         row_labels = species_ordered$Genus,
                         #annotation_col = annotation_col, 
                         annotation_row = annotation_row, 
                         annotation_colors = annotation_colors,
                         cluster_cols = FALSE,
                         cluster_rows = TRUE,
                         main = "Phytohormones",
                         fontsize_row = 8,
                         fontsize_col = 8)
#view
lfc_heatmap

#save by hand with dimensions: 784x634 px 

```

Save data 
```{r}

ancom.p <- ancom$cond_phytohormones #ancom results
lfc  #lfc values  
asvs <- sig_asvs$cond_phytohormones$ASV #vector with ASV names 
dif_abun_stress #contains whether TRUE or FALSE for JA and SA 
log_corr_abn <- log_corr_abn_sig$cond_phytohormones #corrected abundances of DA ASVs
ps_DA #ps with info

#make it into a list  
phyto_dataDA <- list (ancom_test=ancom.p, lfc=lfc, sig_asvs=asvs,
                    binary_stress=dif_abun_stress, log_corr_abn=log_corr_abn,
                    ps=ps_DA)

#save it
save(phyto_dataDA, file="./Results/03_differential_abundance/Phyto_dataDA_16S.Rdata")
```


###Insects
```{r}

#I need:
lfc_sig$cond_real_insects #values of lfc 
dif_abun_stress_sig$cond_real_insects #info if DA in JA and/or SA
ps_16S_split$cond_real_insects #phyloseq

#change name to run the same code 
lfc <- lfc_sig$cond_real_insects
dif_abun_stress <- dif_abun_stress_sig$cond_real_insects

#let's plot only the LFC 

#Now, let's see what happens when the LFC results from ANCOMBC are plotted in a heatmap
rownames(lfc) <- lfc$taxon
lfc <- lfc[,3:4]

#make a matrix from the lfc differences
matrix.log.insects <- as.matrix(lfc) 

#make new ps with only the DA asvs
ps_DA <- phyloseq::prune_taxa(sig_asvs$cond_real_insects$ASV, ps_16S_split$cond_real_insects)

#get species name if available, if not replace with name of ASV
df_row <- as.data.frame((tax_table(ps_DA))[, c("Genus")])
df_row <- df_row %>%
           mutate(asv_name=rownames(.)) #new column with rownames
df_row$Genus <- ifelse(is.na(df_row$Genus), df_row$asv_name, df_row$Genus)
df_row <- df_row %>%
           dplyr::select(Genus)

#order data so it matches
dif_abun_stress_ordered <- dif_abun_stress[order(match(rownames(dif_abun_stress),
                                                     rownames(tax_table(ps_DA)))), , drop=FALSE] 

matrix.log.insects_ordered <- matrix.log.insects[order(match(rownames(matrix.log.insects),
                                                     rownames(tax_table(ps_DA)))), , drop=FALSE] 

#define annotation for rows
annotation_row = data.frame(
    Phylum = as.factor(tax_table(ps_DA)[, "Phylum"]), #to add phylum in the heatmap as well
    Contrast = dif_abun_stress_ordered) 
colnames(annotation_row)[2:3] <- c("JAvsCtrl", "SAvsCtrl")

#change it to factor so it can be colored
annotation_row$JAvsCtrl <- as.factor(annotation_row$JAvsCtrl)
annotation_row$SAvsCtrl <- as.factor(annotation_row$SAvsCtrl)

#check if rownames are in same order 
rownames(tax_table(ps_DA)) #annotation row
rownames(df_row) #yes

#check if names are the same 
rownames(matrix.log.phyto_ordered)
rownames(annotation_row) #yes

#make labels for rows 
df_row
species_ordered <- df_row[order(match(rownames(df_row), rownames(matrix.log.insects_ordered))), , drop = FALSE]

# select colors for factors
#make colors for phylum with custom function
colourCount <- length(levels(annotation_row$Phylum))
getPalette <- colorRampPalette(brewer.pal(colourCount, "Set3"))

#apply to phylum
phylum_col = getPalette(colourCount)
names(phylum_col) = levels(annotation_row$Phylum)
names(color_experiment) = c("cond_phytohormones", "real_insects")

#put all colors together
annotation_colors = list(
     Phylum = phylum_col,
     JAvsCtrl = c("TRUE" = "#E69F00", "FALSE" = "white"),
     SAvsCtrl = c("TRUE" = "#4472C4", "FALSE" = "white"),
     Experiment = color_experiment)

#make colors for matrix
#Select colors for the ends of the values and the middle 
function_color_matrix <- circlize::colorRamp2(c(-2.5, 0, 2.5),
                          c("#0088c9",  "white", "#c85200"))
                         # c("#0000FF", "white", "#FF0000")) #palette blue/red
                         # c("#018571", "white", "#A6611A")) #palette teal/brown

function_color_matrix(seq(-3,3)) #gives you 9 colors with white in the middle 

# now plot the pheatmap
lfc_heatmap <- ComplexHeatmap::pheatmap(matrix.log.insects_ordered,
                         color= function_color_matrix,
                         name="LFC",
                         scale= "none", 
                         row_labels = species_ordered$Genus,
                         #annotation_col = annotation_col, 
                         annotation_row = annotation_row, 
                         annotation_colors = annotation_colors,
                         cluster_cols = FALSE,
                         cluster_rows = TRUE,
                         main = "Insect Herbivory",
                         fontsize_row = 8,
                         fontsize_col = 8)
#view
lfc_heatmap

#save by hand with dimensions: 784x634 px 

```

Save data 
```{r}

ancom.i <- ancom$cond_real_insects #ancom results
lfc  #lfc values  
asvs <- sig_asvs$cond_real_insects$ASV #vector with ASV names 
dif_abun_stress #contains whether TRUE or FALSE for JA and SA 
log_corr_abn <- log_corr_abn_sig$cond_real_insects #corrected abundances of DA ASVs
ps_DA #ps with info

#make it into a list  
insects_dataDA <- list (ancom_test=ancom.p, lfc=lfc, sig_asvs=asvs,
                    binary_stress=dif_abun_stress, log_corr_abn=log_corr_abn,
                    ps=ps_DA)

#save it
save(insects_dataDA, file="./Results/03_differential_abundance/Insects_dataDA_16S.Rdata")

```


###EPG
```{r}

#I need:
lfc_sig$EPG #values of lfc 
dif_abun_stress_sig$EPG #info if DA in JA and/or SA
ps_16S_split$EPG #phyloseq

#change name to run the same code 
lfc <- lfc_sig$EPG
dif_abun_stress <- dif_abun_stress_sig$EPG

#let's plot only the LFC 

#Now, let's see what happens when the LFC results from ANCOMBC are plotted in a heatmap
rownames(lfc) <- lfc$taxon
lfc <- lfc[,3:4]

#make a matrix from the lfc differences
matrix.log.EPG <- as.matrix(lfc) 

#make new ps with only the DA asvs
ps_DA <- phyloseq::prune_taxa(sig_asvs$EPG$ASV, ps_16S_split$EPG)

#get species name if available, if not replace with name of ASV
df_row <- as.data.frame((tax_table(ps_DA))[, c("Genus")])
df_row <- df_row %>%
           mutate(asv_name=rownames(.)) #new column with rownames
df_row$Genus <- ifelse(is.na(df_row$Genus), df_row$asv_name, df_row$Genus)
df_row <- df_row %>%
           dplyr::select(Genus)

#order data so it matches
dif_abun_stress_ordered <- dif_abun_stress[order(match(rownames(dif_abun_stress),
                                                     rownames(tax_table(ps_DA)))), , drop=FALSE] 

matrix.log.EPG_ordered <- matrix.log.EPG[order(match(rownames(matrix.log.EPG),
                                                     rownames(tax_table(ps_DA)))), , drop=FALSE] 

#define annotation for rows
annotation_row = data.frame(
    Phylum = as.factor(tax_table(ps_DA)[, "Phylum"]), #to add phylum in the heatmap as well
    Contrast = dif_abun_stress_ordered) 
colnames(annotation_row)[2:3] <- c("JAvsCtrl", "SAvsCtrl")

#change it to factor so it can be colored
annotation_row$JAvsCtrl <- as.factor(annotation_row$JAvsCtrl)
annotation_row$SAvsCtrl <- as.factor(annotation_row$SAvsCtrl)

#check if rownames are in same order 
rownames(tax_table(ps_DA)) #annotation row
rownames(df_row) #yes

#check if names are the same 
rownames(matrix.log.EPG_ordered)
rownames(annotation_row) #yes

#make labels for rows 
df_row
species_ordered <- df_row[order(match(rownames(df_row), rownames(matrix.log.EPG_ordered))), , drop = FALSE]

# select colors for factors
#make colors for phylum with custom function
colourCount <- length(levels(annotation_row$Phylum))
getPalette <- colorRampPalette(brewer.pal(colourCount, "Set3"))

#apply to phylum
phylum_col = getPalette(colourCount)
names(phylum_col) = levels(annotation_row$Phylum)
names(color_experiment) = c("cond_phytohormones", "real_insects")

#put all colors together
annotation_colors = list(
     Phylum = phylum_col,
     JAvsCtrl = c("TRUE" = "#E69F00", "FALSE" = "white"),
     SAvsCtrl = c("TRUE" = "#4472C4", "FALSE" = "white"),
     Experiment = color_experiment)

#make colors for matrix
#Select colors for the ends of the values and the middle 
function_color_matrix <- circlize::colorRamp2(c(-2.5, 0, 2.5),
                          c("#0088c9",  "white", "#c85200"))
                         # c("#0000FF", "white", "#FF0000")) #palette blue/red
                         # c("#018571", "white", "#A6611A")) #palette teal/brown

function_color_matrix(seq(-3,3)) #gives you 9 colors with white in the middle 

# now plot the pheatmap
lfc_heatmap <- ComplexHeatmap::pheatmap(matrix.log.EPG_ordered,
                         color= function_color_matrix,
                         name="LFC",
                         scale= "none", 
                         row_labels = species_ordered$Genus,
                         #annotation_col = annotation_col, 
                         annotation_row = annotation_row, 
                         annotation_colors = annotation_colors,
                         cluster_cols = FALSE,
                         cluster_rows = TRUE,
                         main = "EPG",
                         fontsize_row = 8,
                         fontsize_col = 8)
#view
lfc_heatmap

#save by hand with dimensions: 784x634 px 

```

Save data 
```{r}

ancom.e <- ancom$EPG #ancom results
lfc  #lfc values  
asvs <- sig_asvs$EPG$ASV #vector with ASV names 
dif_abun_stress #contains whether TRUE or FALSE for JA and SA 
log_corr_abn <- log_corr_abn_sig$EPG #corrected abundances of DA ASVs
ps_DA #ps with info

#make it into a list  
EPG_dataDA <- list (ancom_test=ancom.e, lfc=lfc, sig_asvs=asvs,
                    binary_stress=dif_abun_stress, log_corr_abn=log_corr_abn,
                    ps=ps_DA)

#save it
save(EPG_dataDA, file="./Results/03_differential_abundance/EPG_dataDA_ITS.Rdata")

```

Clean
```{r}
#clean
rm.all.but(keep=c("color_experiment", "phyto_dataDA", "insects_dataDA"), keep_functions = F)
gc()

```

##iTOL
```{r pruning ps}

#load full 16S ps, you'll need it to prune the tree from there 
load("./amplicon_data/Phyloseq_objects/ps_16S.RData")

#load phyto data
load("./Results/03_differential_abundance/Phyto_dataDA_16S.Rdata")

#load insects data
load("./Results/03_differential_abundance/Insects_dataDA_16S.Rdata")

#check name of ASVs to prune a ps 
p <- rownames(phyto_dataDA$ps@tax_table)
i <- rownames(insects_dataDA$ps@tax_table)

intersect <- intersect(p,i) #no
asvs <- c(p,i)
asvs <- unique(asvs) #44

#First, let's get the tree by pruning the original ps_16 with DA ASVs
#make ps with DA asvs
ps_DA <- prune_taxa(asvs, ps_16S) #58 taxa, ok!
Bac_tree <- ps_DA@phy_tree
plot(Bac_tree)
Bac_tree$tip.label

#save it
ape::write.tree(Bac_tree, "./Results/03_differential_abundance/16S_iTOL/16S_DA_tree.tre")

#Now, let's get the sequences to make a new tree as the default one with DADA2 is not great
#get the sequences and save them as FASTA format
ps_DA %>%
refseq() %>%
Biostrings::writeXStringSet("./Results/03_differential_abundance/16S_iTOL/selected_asvs.fna",
             append=FALSE,compress=FALSE,
             compression_level=NA, format="fasta")


#4. Load phylogenetic tree 
new_tree <- read.tree("./Results/03_differential_abundance/16S_iTOL/16S_DA_tree_IQTREE.tre")
new_tree <- ape::multi2di(new_tree) #root it
plot(new_tree)

#check
new_tree

#change tree in the ps object
#copy ps 
ps_DA_tree <- ps_DA

#change it for new
ps_DA_tree@phy_tree <- new_tree

#compare it
ps_DA@phy_tree
ps_DA_tree@phy_tree #this new tree is rooted 

#save it
ape::write.tree(new_tree, "./Results/03_differential_abundance/16S_iTOL/16S_DA_tree_IQTREE_rooted.tre")

#call it again
Bac_tree <- ps_DA@phy_tree

#clean
rm(ps_DA_tree)


```

Now, create a complete dataframe with all the info

```{r dataframe}

##let's get D1:Experiment
df.p <- as.data.frame(p) %>% 
         mutate(Phytohormones= "phytohormones")%>%
         mutate(Experiment="phytohormones")  
colnames(df.p)[1] <-  "id"
  
df.i <- as.data.frame(i)%>% 
         mutate(Insects= "insects")%>%
         mutate(Experiment="insects") 
colnames(df.i)[1] <-  "id"

#merge
df_experiment <- full_join(df.p, df.i)

#get order
order_asvs <- df_experiment$id

#Now, let's get D2:LFC
#get LFC values from both experiments in a single df 
lfc_phyto <- phyto_dataDA$lfc %>%  mutate(Phytohormones= "phytohormones") %>% 
                                   rownames_to_column(var="id")
colnames(lfc_phyto) <- c("id", "lfc_JA_phyto", "lfc_SA_phyto", "Phytohormones")                              

lfc_insects <- insects_dataDA$lfc %>%  mutate(Insects= "insects") %>% 
                                   rownames_to_column(var="id")
colnames(lfc_insects) <- c("id", "lfc_JA_insects", "lfc_SA_insects", "Insects")  

#merge
df_lfc <- full_join(lfc_phyto, lfc_insects)

#Next, let's get D3:Treatment
#phyto
bin_phyto <- phyto_dataDA$binary_stress %>% rownames_to_column(var="id") %>% mutate(Phytohormones="phytohormones")
colnames(bin_phyto) <- c("id","DA_JA_phyto", "DA_SA_phyto", "Phytohormones")
bin_phyto <- bin_phyto %>% mutate(DA_JA_phyto = case_when(DA_JA_phyto == TRUE ~ "YES",
                                DA_JA_phyto == FALSE ~ "NO")) %>% 
                              mutate(DA_SA_phyto = case_when(DA_SA_phyto == TRUE ~ "YES",
                                DA_SA_phyto == FALSE ~ "NO"))
bin_phyto <- bin_phyto %>%
                mutate(DA_phyto = case_when(
                  DA_JA_phyto == "YES" & DA_SA_phyto == "NO" ~ "JA_Only",
                  DA_JA_phyto == "NO" & DA_SA_phyto == "YES" ~ "SA_Only",
                  DA_JA_phyto == "YES" & DA_SA_phyto == "YES" ~ "JA_and_SA")) #works

#insects
bin_insects <- insects_dataDA$binary_stress %>% rownames_to_column(var="id") %>% mutate(Insects="insects")
colnames(bin_insects) <- c("id","DA_JA_insects", "DA_SA_insects", "Insects")
bin_insects <- bin_insects %>% mutate(DA_JA_insects = case_when(DA_JA_insects == TRUE ~ "YES",
                                DA_JA_insects == FALSE ~ "NO")) %>% 
                              mutate(DA_SA_insects = case_when(DA_SA_insects == TRUE ~ "YES",
                                DA_SA_insects == FALSE ~ "NO"))
bin_insects <- bin_insects %>%
                mutate(DA_insects = case_when(
                  DA_JA_insects == "YES" & DA_SA_insects == "NO" ~ "JA_Only",
                  DA_JA_insects == "NO" & DA_SA_insects == "YES" ~ "SA_Only",
                  DA_JA_insects == "YES" & DA_SA_insects == "YES" ~ "JA_and_SA")) #works
  
#merge
binary_stress <- full_join(bin_phyto, bin_insects)

#merge to make it complete
df_complete <- full_join(df_lfc, binary_stress)

#add column with Experiment including Phyto, Insects and both
df_complete <- df_complete %>% mutate(Experiment=case_when(
  Phytohormones == "phytohormones" ~ "phytohormones",
  Insects == "insects" ~ "insects"))

#re-order
df_complete <- df_complete %>% relocate(id, Experiment, Phytohormones, Insects)

```


```{r dataframe-taxonomy}

#Finally, let's get taxonomy
#clean it first
tax_table(ps_DA)[, colnames(tax_table(ps_DA))] <- gsub(tax_table(ps_DA)[, colnames(tax_table(ps_DA))],     pattern = "[a-z]__", replacement = "")
#make it a df 
tax_table <- as.data.frame(ps_DA@tax_table[,c(1:7)]) #select the interesting ones 
tax_table$id <- rownames(tax_table) #add id column for join 

#merge
df_complete2 <- left_join(df_complete, tax_table)

#Now this df is complete (for now)

#clean the phyloseq object 
rm(ps_16S)
gc()

```

Move it into iTOL format

iTOL has a specific format to add data into the tree, it has to be in a .txt file and in a certain way coded. The package itol.toolkit helps to make an annotation file directly in R. Once the annotation file (AF) is ready you can import it into the iTOL annotator to add metadata to your phylogenetic tree.


```{r color management}

#color management

#Set-up colors
color_JA <- "#E69F00"
color_SA <- "#4472C4"
color_phyto <- "#CC3399"
color_insects <- "#70AD47"

#Built in colors
display.brewer.all() #gives you the palettes from RColorBrewer


```

```{r Experiment}

#create Annotation files with using the itol.toolkit package

#Experiment: together
temp <- df_complete2%>%
        dplyr::select(c(id,Experiment))

#make df for itol
AF1 <- itol.toolkit::create_unit(data = temp, 
                         key = "AF1_Experiment", 
                         type = "DATASET_COLORSTRIP",
                         tree = Bac_tree)

#change colors and parameters
AF1@specific_themes$strip_label$width <- 30 #width of strip
AF1@specific_themes$strip_label$color_branches <- 0 #no color branches
AF1@data$tip <- AF1@data$tip %>%  mutate(`AF1_Experiment$COLOR`=case_when( #change colors 
                                `AF1_Experiment$LABEL`=="phytohormones" ~ color_phyto,
                                `AF1_Experiment$LABEL`=="insects" ~ color_insects))


AF1@common_themes$legend$title <- "Experiment"
AF1@common_themes$legend$horizontal <- 0
AF1@common_themes$legend$labels <- c("Herbivore-induced","Phytohormone-induced")

#save
itol.toolkit::write_unit(AF1, "./Results/03_differential_abundance/16S_iTOL")


```

Treatment
```{r Treatment}
## Treatment
df_Treatment <- df_complete2%>%
                dplyr::select(id,DA_JA_phyto,DA_JA_insects, DA_SA_phyto, DA_SA_insects)%>%
                relocate(id)
#replace NA's
df_Treatment$DA_JA_phyto <- df_Treatment$DA_JA_phyto %>% replace_na('NO')
df_Treatment$DA_SA_phyto <- df_Treatment$DA_SA_phyto %>% replace_na('NO')
df_Treatment$DA_JA_insects <- df_Treatment$DA_JA_insects %>% replace_na('NO')
df_Treatment$DA_SA_insects <- df_Treatment$DA_SA_insects %>% replace_na('NO')

#replace 0's for -1 and Yes for 1 so iTOL recognizes full and empty symbols
df_Treatment2 <- df_Treatment %>% 
                mutate(DA_JA_phyto=case_when(DA_JA_phyto== "YES" ~ "1", TRUE ~ "-1")) %>% 
                mutate(DA_SA_phyto=case_when(DA_SA_phyto== "YES" ~ "1", TRUE ~ "-1")) %>%
                mutate(DA_JA_insects=case_when(DA_JA_insects== "YES" ~ "1", TRUE ~ "-1")) %>%
                mutate(DA_SA_insects=case_when(DA_SA_insects== "YES" ~ "1", TRUE ~ "-1"))

## By Experiment
#Phyto
Phyto <- df_Treatment2 %>% dplyr::select(id, DA_JA_phyto, DA_SA_phyto)
AF2.1 <- itol.toolkit::create_unit(data = Phyto, 
                         key = "AF2.1_Treatment_Phyto", 
                         type = "DATASET_BINARY",
                         tree = Bac_tree)

AF2.1@field$labels <- c("JA_phyto",  "SA_phyto")
AF2.1@field$colors <- c("#E69F00",  "#4472C4")

itol.toolkit::write_unit(AF2.1, "./Results/03_differential_abundance/16S_iTOL")

#Insects
insects <- df_Treatment2 %>% dplyr::select(id, DA_JA_insects, DA_SA_insects)
AF2.2 <- itol.toolkit::create_unit(data = insects, 
                         key = "AF2.2_Treatment_insects", 
                         type = "DATASET_BINARY",
                         tree = Bac_tree)

AF2.2@field$labels <- c("JA_insects",  "SA_insects")
AF2.2@field$colors <- c("#E69F00",  "#4472C4")

itol.toolkit::write_unit(AF2.2, "./Results/03_differential_abundance/16S_iTOL")

```

Independently
```{r Treatment-independently}

##JA
JA_phyto <- df_Treatment2 %>% dplyr::select(id, DA_JA_phyto)
AF3.1 <- itol.toolkit::create_unit(data = JA_phyto, 
                         key = "AF3.1_Treatment_JA_Phyto", 
                         type = "DATASET_BINARY",
                         tree = Bac_tree)

AF3.1@field$labels <- "JA_phyto"
AF3.1@field$colors <- "#E69F00"

itol.toolkit::write_unit(AF3.1, "./Results/03_differential_abundance/16S_iTOL")

JA_insects <- df_Treatment2 %>% dplyr::select(id, DA_JA_insects)
AF3.2 <- itol.toolkit::create_unit(data = JA_insects, 
                         key = "AF3.1_Treatment_JA_insects", 
                         type = "DATASET_BINARY",
                         tree = Bac_tree)

AF3.2@field$labels <- "JA_insects"
AF3.2@field$colors <- "#E69F00"

itol.toolkit::write_unit(AF3.2, "./Results/03_differential_abundance/16S_iTOL")

##SA
SA_phyto <- df_Treatment2 %>% dplyr::select(id, DA_SA_phyto)
AF3.3 <- itol.toolkit::create_unit(data = SA_phyto, 
                         key = "AF3.3_Treatment_SA_Phyto", 
                         type = "DATASET_BINARY",
                         tree = Bac_tree)

AF3.3@field$labels <- "SA_phyto"
AF3.3@field$colors <- "#4472C4"

itol.toolkit::write_unit(AF3.3, "./Results/03_differential_abundance/16S_iTOL")

SA_insects <- df_Treatment2 %>% dplyr::select(id, DA_SA_insects)
AF3.4 <- itol.toolkit::create_unit(data = SA_insects, 
                         key = "AF3.4_Treatment_SA_insects", 
                         type = "DATASET_BINARY",
                         tree = Bac_tree)

AF3.4@field$labels <- "SA_insects"
AF3.4@field$colors <- "#4472C4"

itol.toolkit::write_unit(AF3.4, "./Results/03_differential_abundance/16S_iTOL")

```

LFC
```{r LFC-heatmap}

## LFC dataset
df_LFC <- df_complete2%>%
                dplyr::select(id,lfc_JA_phyto, lfc_JA_insects, lfc_SA_phyto, lfc_SA_insects)%>%
                relocate(id)%>%
                as.tibble()
#let's round up
df_LFC <- df_LFC %>% 
          dplyr::mutate(across(where(is.numeric), round, 2))

#First, lets make a single df 
df_LFC_unique <- full_join(df_Treatment,df_LFC)
#Now let's apply some rules
df_LFC_unique <- df_LFC_unique %>%
                  mutate(lfc_JA_phyto=
                           case_when(DA_JA_phyto == "NO" ~ NA, TRUE ~ lfc_JA_phyto))%>% 
                  mutate(lfc_SA_phyto=
                           case_when(DA_SA_phyto == "NO" ~ NA, TRUE ~ lfc_SA_phyto))%>% 
                  mutate(lfc_JA_insects=
                           case_when(DA_JA_insects == "NO" ~ NA, TRUE ~ lfc_JA_insects))%>% 
                  mutate(lfc_SA_insects=
                           case_when(DA_SA_insects == "NO" ~ NA, TRUE ~ lfc_SA_insects))

#By experiment
Phyto <-  df_LFC_unique %>% dplyr::select(id, lfc_JA_phyto, lfc_SA_phyto)

#get Annotation File for Heatmap
AF3.1 <- itol.toolkit::create_unit(data = Phyto , 
                         key = "AF3.1_LFC_Heatmap_Phyto", 
                         type = "DATASET_HEATMAP",
                         tree = Bac_tree)  #works

AF3.1@data$tip[AF3.1@data$tip == 0] <- NA

#choose colors
display.brewer.pal(n=8, name='RdBu') #see palette
color_heatmap <- brewer.pal(n = 8, name = "RdBu") #get colors
min <- color_heatmap[8]
max <- color_heatmap[1]

#select colors 
AF3.1@specific_themes$heatmap$color$min <- min
AF3.1@specific_themes$heatmap$color$max <- max
AF3.1@specific_themes$heatmap$color$mid <- "#ebeceb"
AF3.1@specific_themes$heatmap$value$mid <- 0
AF3.1@specific_themes$heatmap$color$nan <- "#FFFFFF"

#change range so the colors mean the same across experiment and treatment
AF3.1@specific_themes$heatmap$value$max <- 2.75
AF3.1@specific_themes$heatmap$value$min <- -2.55
AF3.1@specific_themes$heatmap$value$mid <- 0

AF3.1@specific_themes$heatmap$tree$tree_display <- 0 #no display of tree in heatmap

itol.toolkit::write_unit(AF3.1, "./Results/03_differential_abundance/16S_iTOL") #works

#By experiment
Insects <-  df_LFC_unique %>% dplyr::select(id, lfc_JA_insects, lfc_SA_insects)

#get Annotation File for Heatmap
AF3.2 <- itol.toolkit::create_unit(data = Insects , 
                         key = "AF3.2_LFC_Heatmap_Insects", 
                         type = "DATASET_HEATMAP",
                         tree = Bac_tree)  #works

AF3.2@data$tip[AF3.2@data$tip == 0] <- NA

#select colors 
AF3.2@specific_themes$heatmap$color$min <- min
AF3.2@specific_themes$heatmap$color$max <- max
AF3.2@specific_themes$heatmap$color$mid <- "#ebeceb"
AF3.2@specific_themes$heatmap$value$mid <- 0
AF3.2@specific_themes$heatmap$color$nan <- "#FFFFFF"

#change range so the colors mean the same across experiment and treatment
AF3.2@specific_themes$heatmap$value$max <- 2.75
AF3.2@specific_themes$heatmap$value$min <- -2.55
AF3.2@specific_themes$heatmap$value$mid <- 0

AF3.2@specific_themes$heatmap$tree$tree_display <- 0 #no display of tree in heatmap

itol.toolkit::write_unit(AF3.2, "./Results/03_differential_abundance/16S_iTOL") #works

```

Now I need to add the taxonomy and change the labels

```{r Taxonomy}

#first, let's change it for Family labels
#Family
df_labels_F <- df_complete2 %>%
               dplyr::select(id, Family)

AF4 <- create_unit(data = df_labels_F,
                       key = "AF4_label_Family",
                       type = "LABELS",
                       tree = Bac_tree)

itol.toolkit::write_unit(AF4, "./Results/03_differential_abundance/16S_iTOL")


#Genus
df_labels_G <- df_complete2 %>%
               dplyr::select(id, Genus)

#replace Genus names with names of ASV and add number to Genus so its clear that it's different ASVs
df_labels_G <- df_labels_G %>% mutate(number=str_split_i(df_labels_G$id, "_", 2))%>% #get number of ASV
                                mutate(Genus=paste(.$Genus, .$number, sep= "-"))%>% #make new with number and genus
                                mutate(Genus=gsub("NA", "fASV", .$Genus))%>%
                                dplyr::select(-number) #ok
                                      
AF5 <- create_unit(data = df_labels_G,
                       key = "AF5_label_Genus",
                       type = "LABELS",
                       tree = Bac_tree)

itol.toolkit::write_unit(AF5, "./Results/03_differential_abundance/16S_iTOL")


#Make a range df for Phylum now
df_Phylum <- df_complete2 %>%
               dplyr::select(id, Phylum, Class)

df_Phylum <- df_Phylum %>% mutate(Phylum=case_when(
                          Class==" Alphaproteobacteria" ~ " Alphaproteobacteria",
                          Class == " Gammaproteobacteria" ~ " Gammaproteobacteria",
                          TRUE ~ Phylum))

#make a vector with colors 
Phylum <- unique(df_Phylum$Phylum) #11
Phylum <- sort(Phylum)
color_phylum <- c("#FDB462","#8DD3C7","#FFFFB3", "#FB8072","#80B1D3","#8DD3D2",
                  "#FCCDE5", "#BC80BD","#BEBADA", "#B3DE69", "#CCEBC5")

#make it a df
df_phyla_col <- data.frame(Phylum, color_phylum)

#merge with df_Phylum
df_Phylum <- df_Phylum%>%
             dplyr::select(-Class) %>% 
             dplyr::inner_join(df_phyla_col)

AF6 <- create_unit(data = df_Phylum, 
                      key = "AF6_Phylum_colors", 
                      type = "TREE_COLORS", 
                      subtype = "range",
                      tree = Bac_tree)

itol.toolkit::write_unit(AF6, "./Results/03_differential_abundance/16S_iTOL")
```

Save for later
```{r}

#Save environment
save.image(file="./R_Environments/03_differential_abundance/16S_iTOL.Rdata")

#clean
rm.all.but(keep=c("color_JA", "color_SA"), keep_functions = F)
gc()

```


#ITS

```{r}
#Load ITS data

#ITS
load(file = "./amplicon_data/Phyloseq_objects/ps_ITS_split.RData")
ps_ITS_split

```


##Pre-filtering

```{r}
#First, make a pre-filtering of those taxa that are low in prevalence (<20%)

ps_ITS_filt <- lapply(ps_ITS_split, function(ps){
  metagMisc::phyloseq_filter_prevalence(ps, prev.trh = 0.10, #at least 20% of samples
                                          abund.trh = NULL) #no filtering by abundance
})

#check
ps_ITS_filt #less than 150 taxa, better to use 10% prevalence filter 

#clean
rm(ps_ITS_filt)

```

##ANCOM
```{r}

ancom <- lapply(ps_ITS_split, function(ps){
  set.seed(123)
  ANCOMBC::ancombc(phyloseq = ps, 
                      formula = "plant_pathway_induced", 
                      p_adj_method = "fdr", #same as "BH" Benjamini & Hochberg
                      prv_cut = 0.10,
                      group = "plant_pathway_induced", 
                      struc_zero = TRUE, #with structural zero's
                      neg_lb = FALSE, #classifies structural zero's if it's zero and not its asymptotic lower bound 
                      tol = 1e-5,               
                      max_iter = 100, 
                      conserve = TRUE, #conservative variance of the test statistics 
                      alpha = 0.05,
                      verbose=TRUE,
                      global=FALSE)})

save(ancom, file = "./Results/03_differential_abundance/ancom_ITS.Rdata")

```

Get results 
```{r}

head(ancom$cond_phytohormones$feature_table)

#check
lapply(ancom, function(ancom_result){head(ancom_result$res$lfc)})#LFC
lapply(ancom, function(ancom_result){head(ancom_result$res$se)})#standard error
lapply(ancom, function(ancom_result){head(ancom_result$res$W)}) #test statistic LFC/se

#get sample fractions
samp_frac <- lapply(ancom, function(ancom_result){ancom_result$samp_frac})#ancom$samp_frac

# Get log transformed counts and add pseudo-count (1) to feature table to avoid taking the log of 0
log_obs_abn <-lapply(ancom, function(ancom_result){log(ancom_result$feature_table +1)})

# Adjust the log observed abundances with the sample fraction
log_corr_abn <-mapply(function(x,y){
                     t(t(x) - y)},
                     x=log_obs_abn,
                     y=samp_frac)
  
#Results
log_corr_abn #bias-corrected log abundances
lfc <- lapply(ancom, function(ancom_result){ancom_result$res$lfc}) #log-fold change according to control 
se_lfc <- lapply(ancom, function(ancom_result){ancom_result$res$se}) #standard error of log-fold change

#get significant ASVs   
JA_asvs <- lapply(ancom, function(ancom_result){
           asvs <- ancom_result$res$diff_abn %>%
            dplyr::filter(plant_pathway_inducedJA == TRUE) #26,12,8
           asvs$ASV <- asvs$taxon
           return(asvs)})

SA_asvs <- lapply(ancom, function(ancom_result){
           asvs <- ancom_result$res$diff_abn %>%
            dplyr::filter(plant_pathway_inducedSA == TRUE) #9,4,10
           asvs$ASV <- asvs$taxon
           return(asvs)})

#Now I need to get the unique names 
sig_asvs <- mapply(function(x,y){
  c(x$ASV, y$ASV)},
  x=JA_asvs,
  y=SA_asvs)

#make it into a dataframe format
sig_asvs <- lapply(sig_asvs, function(x){
   unique <- unique(x)
   unique <- as.data.frame(unique)
   colnames(unique) <- "ASV"
   return(unique)})  

#With this I can slice the different data frames to plot them

#1.bias-corrected abundances
log_corr_abn_sig <- mapply(function(x,y){x[y$ASV,]},
  x=log_corr_abn,
  y=sig_asvs) #works because its a matrix 

#2.log-fold change vs controls
lfc_sig <- mapply(function(x,y){x[x$taxon %in% y$ASV,]},
                  x=lfc,
                  y=sig_asvs,
                  SIMPLIFY = FALSE) #to keep it as a list and not a single thing

#3.Differential abundance according to stress 
dif_abun_stress_sig <- mapply(function(x,y){x$res$diff_abn[x$res$diff_abn$taxon %in% y$ASV,]},
                           x=ancom,
                           y=sig_asvs,
                           SIMPLIFY = FALSE) 

#place ASV as rowname to avoid mistakes later
dif_abun_stress_sig <- lapply(dif_abun_stress_sig, function(df){
  rownames(df) <- df$taxon
  df <- df[,3:4]
  return(df)
})



```

##Euler diagrams
Check it with an Euler diagram
```{r}

library(eulerr)

#First, arrange the data
JA_asvs$cond_phytohormones$ASV
SA_asvs$cond_phytohormones$ASV


phyto <- list(
           JA = JA_asvs$cond_phytohormones$ASV,
           SA = SA_asvs$cond_phytohormones$ASV)

insects <- list(
           JA = JA_asvs$cond_real_insects$ASV,
           SA = SA_asvs$cond_real_insects$ASV)

EPG <- list(
           JA = JA_asvs$EPG$ASV,
           SA = SA_asvs$EPG$ASV)

da_list <- list(phyto=phyto, insects=insects, EPG=EPG)

#calculate overlapp
venn_df <- lapply(da_list, function(list){
  calculate.overlap(list)})
  
#get numbers 
euler <- lapply(venn_df, function(list){
  #get length
  JA <- length(list$a1)
  SA <- length(list$a2)
  JA_SA <- length(list$a3)
  
  list <- list(JA=JA, SA=SA, JA_SA=JA_SA)
  return(list)
})

#plot it
venn_phyto <- draw.pairwise.venn(area1=euler$phyto$JA,
                                 area2=euler$phyto$SA,
                                 cross.area=euler$phyto$JA_SA,
                   euler.d = TRUE,
                   scaled = TRUE,
                   ext.text = FALSE,
                   #category=c("JA", "SA"),
                   lty = 1, #outline
                   col = c(color_JA, color_SA),
                   lwd=4,
                   #ext.line.lwd = 0.5,
                   fill = c(color_JA, color_SA), #color of area
                   cex=1.5, #size of labels
                   alpha = rep(0.8, 2), #transparency of color
                   cat.pos = 0,
                   cat.dist = 0,
                   fontfamily = fonts$sans,
                   cat.fontfamily = fonts$sans,
                   inverted = FALSE,
                   print.mode = c("raw", "percent"),
                   rotation.degree = 0)

#save it by hand 500x500 px

venn_insects <- draw.pairwise.venn(area1=euler$insects$JA,
                                 area2=euler$insects$SA,
                                 cross.area=euler$insects$JA_SA,
                   euler.d = TRUE,
                   scaled = TRUE,
                   ext.text = FALSE,
                   #category=c("JA", "SA"),
                   lty = 1, #outline
                   col = c(color_JA, color_SA),
                   lwd=4,
                   #ext.line.lwd = 0.5,
                   fill = c(color_JA, color_SA), #color of area
                   cex=1.5, #size of labels
                   alpha = rep(0.8, 2), #transparency of color
                   cat.pos = 0,
                   cat.dist = 0,
                   fontfamily = fonts$sans,
                   cat.fontfamily = fonts$sans,
                   inverted = FALSE,
                   print.mode = c("raw", "percent"),
                   rotation.degree = 0)


venn_EPG <- draw.pairwise.venn(area1=euler$EPG$JA,
                                 area2=euler$EPG$SA,
                                 cross.area=euler$EPG$JA_SA,
                   euler.d = TRUE,
                   scaled = TRUE,
                   ext.text = FALSE,
                   #category=c("JA", "SA"),
                   lty = 1, #outline
                   col = c(color_JA, color_SA),
                   lwd=4,
                   #ext.line.lwd = 0.5,
                   fill = c(color_JA, color_SA), #color of area
                   cex=1.5, #size of labels
                   alpha = rep(0.8, 2), #transparency of color
                   cat.pos = 0,
                   cat.dist = 0,
                   fontfamily = fonts$sans,
                   cat.fontfamily = fonts$sans,
                   inverted = FALSE,
                   print.mode = c("raw", "percent"),
                   rotation.degree = 0)

#clean
rm(venn_phyto, venn_insects, venn_EPG, euler, venn_df)

```

##Heatmap LFC by treatment

To make my life easier I'll plot one per experiment and just run the same code 

###Phyto


```{r}

#I need:
lfc_sig$cond_phytohormones #values of lfc 
dif_abun_stress_sig$cond_phytohormones #info if DA in JA and/or SA
ps_ITS_split$cond_phytohormones #phyloseq

#change name to run the same code 
lfc <- lfc_sig$cond_phytohormones
dif_abun_stress <- dif_abun_stress_sig$cond_phytohormones

#let's plot only the LFC 

#Now, let's see what happens when the LFC results from ANCOMBC are plotted in a heatmap
rownames(lfc) <- lfc$taxon
lfc <- lfc[,3:4]

#make a matrix from the lfc differences
matrix.log.phyto <- as.matrix(lfc) 

#make new ps with only the DA asvs
ps_DA <- phyloseq::prune_taxa(sig_asvs$cond_phytohormones$ASV, ps_ITS_split$cond_phytohormones)

#get species name if available, if not replace with name of ASV
df_row <- as.data.frame((tax_table(ps_DA))[, c("Genus")])
df_row <- df_row %>%
           mutate(asv_name=rownames(.)) #new column with rownames
df_row$Genus <- ifelse(is.na(df_row$Genus), df_row$asv_name, df_row$Genus)
df_row <- df_row %>%
           dplyr::select(Genus)

#get life-guild trait if available
df_guild <- as.data.frame((tax_table(ps_DA))[,c("Trophic.Mode")])
df_guild <- df_guild %>% 
            mutate(asv_name=rownames(.))
df_guild$Trophic.Mode <- ifelse(df_guild$Trophic.Mode == "-", 'unknown', df_guild$Trophic.Mode)

#order data so it matches
dif_abun_stress_ordered <- dif_abun_stress[order(match(rownames(dif_abun_stress),
                                                     rownames(tax_table(ps_DA)))), , drop=FALSE] 

matrix.log.phyto_ordered <- matrix.log.phyto[order(match(rownames(matrix.log.phyto),
                                                     rownames(tax_table(ps_DA)))), , drop=FALSE] 

#define annotation for rows
annotation_row = data.frame(
    Phylum = as.factor(tax_table(ps_DA)[, "Phylum"]), #to add phylum in the heatmap as well
    Guild = as.factor(tax_table(ps_DA)[, "Trophic.Mode"]),
    Contrast = dif_abun_stress_ordered) 
colnames(annotation_row)[3:4] <- c("JAvsCtrl", "SAvsCtrl")

#change it to factor so it can be colored
annotation_row$JAvsCtrl <- as.factor(annotation_row$JAvsCtrl)
annotation_row$SAvsCtrl <- as.factor(annotation_row$SAvsCtrl)

#check if rownames are in same order 
rownames(tax_table(ps_DA)) #annotation row
rownames(df_row) #yes

#check if names are the same 
rownames(matrix.log.phyto_ordered)
rownames(annotation_row) #yes

#make labels for rows 
df_row
species_ordered <- df_row[order(match(rownames(df_row), rownames(matrix.log.phyto_ordered))), , drop = FALSE]

# select colors for factors
#make colors for phylum with custom function
colourCount <- length(levels(annotation_row$Phylum))
getPalette <- colorRampPalette(brewer.pal(colourCount, "Set3"))

#apply to phylum
phylum_col = getPalette(colourCount)
names(phylum_col) = levels(annotation_row$Phylum)

#apply to guild
colourCount <- length(levels(annotation_row$Guild))
getPalette <- colorRampPalette(brewer.pal(colourCount, "Paired"))
getPalette(colourCount)
guild_colors <- c("#A6CEE3", "#1F78B4", "#B2DF8A", "#33A02C", "#FB9A99", "#E31A1C", "#FDBF6F")
levels(annotation_row$Guild)
guild_colors <- c("gray","#E31A1C", "#FDBF6F", "#A6CEE3", "#1F78B4", "#B2DF8A", "#33A02C")
names(guild_colors) <- levels(annotation_row$Guild)

#change name of experiment so it works
names(color_experiment) = c("cond_phytohormones", "real_insects")

#put all colors together
annotation_colors = list(
     Phylum = phylum_col,
     Guild = guild_colors,
     JAvsCtrl = c("TRUE" = "#E69F00", "FALSE" = "white"),
     SAvsCtrl = c("TRUE" = "#4472C4", "FALSE" = "white"),
     Experiment = color_experiment)

#make colors for matrix
#Select colors for the ends of the values and the middle 
function_color_matrix <- circlize::colorRamp2(c(-2.5, 0, 2.5),
                          c("#0088c9",  "white", "#c85200"))
                         # c("#0000FF", "white", "#FF0000")) #palette blue/red
                         # c("#018571", "white", "#A6611A")) #palette teal/brown

function_color_matrix(seq(-3,3)) #gives you 9 colors with white in the middle 

# now plot the pheatmap
lfc_heatmap <- ComplexHeatmap::pheatmap(matrix.log.phyto_ordered,
                         color= function_color_matrix,
                         name="LFC",
                         scale= "none", 
                         row_labels = species_ordered$Genus,
                         #annotation_col = annotation_col, 
                         annotation_row = annotation_row, 
                         annotation_colors = annotation_colors,
                         cluster_cols = FALSE,
                         cluster_rows = TRUE,
                         main = "Phytohormones-ITS",
                         fontsize_row = 8,
                         fontsize_col = 8)
#view
lfc_heatmap

#save by hand with dimensions: 784x634 px 

```

Save data
```{r}

ancom.p <- ancom$cond_phytohormones #ancom results
lfc  #lfc values  
asvs <- sig_asvs$cond_phytohormones$ASV #vector with ASV names 
dif_abun_stress #contains whether TRUE or FALSE for JA and SA 
log_corr_abn <- log_corr_abn_sig$cond_phytohormones #corrected abundances of DA ASVs
ps_DA #ps with info

#make it into a list  
phyto_dataDA <- list (ancom_test=ancom.p, lfc=lfc, sig_asvs=asvs,
                    binary_stress=dif_abun_stress, log_corr_abn=log_corr_abn,
                    ps=ps_DA)

#save it
save(phyto_dataDA, file="./Results/03_differential_abundance/phyto_dataDA.Rdata")

#clean
rm.all.but(keep=c("insects", "EPG", "ancom", "sig_asvs", "dif_abun_stress_sig", "lfc_sig",
                  "log_corr_abn_sig", "ps_ITS_split" ), keep_functions = F)

```

###Insects
```{r}

#I need:
lfc_sig$cond_real_insects #values of lfc 
dif_abun_stress_sig$cond_real_insects #info if DA in JA and/or SA
ps_ITS_split$cond_real_insects #phyloseq

#change name to run the same code 
lfc <- lfc_sig$cond_real_insects
dif_abun_stress <- dif_abun_stress_sig$cond_real_insects

#let's plot only the LFC 

#Now, let's see what happens when the LFC results from ANCOMBC are plotted in a heatmap
rownames(lfc) <- lfc$taxon
lfc <- lfc[,3:4]

#make a matrix from the lfc differences
matrix.log.insects <- as.matrix(lfc) 

#make new ps with only the DA asvs
ps_DA <- phyloseq::prune_taxa(sig_asvs$cond_real_insects$ASV, ps_ITS_split$cond_real_insects)

#get species name if available, if not replace with name of ASV
df_row <- as.data.frame((tax_table(ps_DA))[, c("Genus")])
df_row <- df_row %>%
           mutate(asv_name=rownames(.)) #new column with rownames
df_row$Genus <- ifelse(is.na(df_row$Genus), df_row$asv_name, df_row$Genus)
df_row <- df_row %>%
           dplyr::select(Genus)

#order data so it matches
dif_abun_stress_ordered <- dif_abun_stress[order(match(rownames(dif_abun_stress),
                                                     rownames(tax_table(ps_DA)))), , drop=FALSE] 

matrix.log.insects_ordered <- matrix.log.insects[order(match(rownames(matrix.log.insects),
                                                     rownames(tax_table(ps_DA)))), , drop=FALSE] 

#define annotation for rows
annotation_row = data.frame(
    Phylum = as.factor(tax_table(ps_DA)[, "Phylum"]), #to add phylum in the heatmap as well
    Guild = as.factor(tax_table(ps_DA)[, "Trophic.Mode"]),
    Contrast = dif_abun_stress_ordered) 
colnames(annotation_row)[3:4] <- c("JAvsCtrl", "SAvsCtrl")

#change it to factor so it can be colored
annotation_row$JAvsCtrl <- as.factor(annotation_row$JAvsCtrl)
annotation_row$SAvsCtrl <- as.factor(annotation_row$SAvsCtrl)

#check if rownames are in same order 
rownames(tax_table(ps_DA)) #annotation row
rownames(df_row) #yes

#check if names are the same 
rownames(matrix.log.insects_ordered)
rownames(annotation_row) #yes

#make labels for rows 
df_row
species_ordered <- df_row[order(match(rownames(df_row), rownames(matrix.log.insects_ordered))), , drop = FALSE]

# select colors for factors
#make colors for phylum with custom function
colourCount <- length(levels(annotation_row$Phylum))
getPalette <- colorRampPalette(brewer.pal(colourCount, "Set3"))

#apply to phylum
phylum_col <- getPalette(colourCount)
phylum_col <- c("#8DD3C7", "#BEBADA")
names(phylum_col) = levels(annotation_row$Phylum)

#apply to guild
colourCount <- length(levels(annotation_row$Guild))
getPalette <- colorRampPalette(brewer.pal(colourCount, "Paired"))
getPalette(colourCount)
guild_colors <- c("#A6CEE3", "#1F78B4", "#B2DF8A", "#33A02C", "#FB9A99", "#E31A1C", "#FDBF6F")
levels(annotation_row$Guild)
guild_colors <- c("gray","#E31A1C", "#FDBF6F", "#A6CEE3", "#1F78B4", "#B2DF8A", "#33A02C")
names(guild_colors) <- levels(annotation_row$Guild)

#change name of experiment so it works
color_experiment <- c("#CC3399", "#70AD47")
names(color_experiment) = c("cond_phytohormones", "real_insects")

#put all colors together
annotation_colors = list(
     Phylum = phylum_col,
     Guild = guild_colors,
     JAvsCtrl = c("TRUE" = "#E69F00", "FALSE" = "white"),
     SAvsCtrl = c("TRUE" = "#4472C4", "FALSE" = "white"),
     Experiment = color_experiment)

#make colors for matrix
#Select colors for the ends of the values and the middle 
function_color_matrix <- circlize::colorRamp2(c(-2.5, 0, 2.5),
                          c("#0088c9",  "white", "#c85200"))

function_color_matrix(seq(-3,3)) #gives you 9 colors with white in the middle 

# now plot the pheatmap
lfc_heatmap <- ComplexHeatmap::pheatmap(matrix.log.insects_ordered,
                         color= function_color_matrix,
                         name="LFC",
                         scale= "none", 
                         row_labels = species_ordered$Genus,
                         #annotation_col = annotation_col, 
                         annotation_row = annotation_row, 
                         annotation_colors = annotation_colors,
                         cluster_cols = FALSE,
                         cluster_rows = TRUE,
                         main = "Insect Herbivory-ITS",
                         fontsize_row = 8,
                         fontsize_col = 8)
#view
lfc_heatmap

#save by hand with dimensions: 784x634 px 

```

Save data
```{r}

ancom.i <- ancom$cond_real_insects #ancom results
lfc  #lfc values  
asvs <- sig_asvs$cond_real_insects$ASV #vector with ASV names 
dif_abun_stress #contains whether TRUE or FALSE for JA and SA 
log_corr_abn <- log_corr_abn_sig$cond_real_insects #corrected abundances of DA ASVs
ps_DA #ps with info

#make it into a list  
insects_dataDA <- list (ancom_test=ancom.i, lfc=lfc, sig_asvs=asvs,
                    binary_stress=dif_abun_stress, log_corr_abn=log_corr_abn,
                    ps=ps_DA)

#save it
save(insects_dataDA, file="./Results/03_differential_abundance/insects_dataDA.Rdata")

#clean
rm.all.but(keep=c("EPG", "ancom", "sig_asvs", "dif_abun_stress_sig", "lfc_sig",
                  "log_corr_abn_sig", "ps_ITS_split"), keep_functions = F)

```


###EPG
```{r}

#I need:
lfc_sig$EPG #values of lfc 
dif_abun_stress_sig$EPG #info if DA in JA and/or SA
ps_ITS_split$EPG #phyloseq

#change name to run the same code 
lfc <- lfc_sig$EPG
dif_abun_stress <- dif_abun_stress_sig$EPG

#let's plot only the LFC 

#Now, let's see what happens when the LFC results from ANCOMBC are plotted in a heatmap
rownames(lfc) <- lfc$taxon
lfc <- lfc[,3:4]

#make a matrix from the lfc differences
matrix.log.EPG <- as.matrix(lfc) 

#make new ps with only the DA asvs
ps_DA <- phyloseq::prune_taxa(sig_asvs$EPG$ASV, ps_ITS_split$EPG)

#get species name if available, if not replace with name of ASV
df_row <- as.data.frame((tax_table(ps_DA))[, c("Genus")])
df_row <- df_row %>%
           mutate(asv_name=rownames(.)) #new column with rownames
df_row$Genus <- ifelse(is.na(df_row$Genus), df_row$asv_name, df_row$Genus)
df_row <- df_row %>%
           dplyr::select(Genus)

#order data so it matches
dif_abun_stress_ordered <- dif_abun_stress[order(match(rownames(dif_abun_stress),
                                                     rownames(tax_table(ps_DA)))), , drop=FALSE] 

matrix.log.EPG_ordered <- matrix.log.EPG[order(match(rownames(matrix.log.EPG),
                                                     rownames(tax_table(ps_DA)))), , drop=FALSE] 

#define annotation for rows
annotation_row = data.frame(
    Phylum = as.factor(tax_table(ps_DA)[, "Phylum"]), #to add phylum in the heatmap as well
    Guild = as.factor(tax_table(ps_DA)[, "Trophic.Mode"]),
    Contrast = dif_abun_stress_ordered) 
colnames(annotation_row)[3:4] <- c("JAvsCtrl", "SAvsCtrl")

#change it to factor so it can be colored
annotation_row$JAvsCtrl <- as.factor(annotation_row$JAvsCtrl)
annotation_row$SAvsCtrl <- as.factor(annotation_row$SAvsCtrl)

#check if rownames are in same order 
rownames(tax_table(ps_DA)) #annotation row
rownames(df_row) #yes

#check if names are the same 
rownames(matrix.log.EPG_ordered)
rownames(annotation_row) #yes

#make labels for rows 
df_row
species_ordered <- df_row[order(match(rownames(df_row), rownames(matrix.log.EPG_ordered))), , drop = FALSE]

# select colors for factors
#make colors for phylum with custom function
colourCount <- length(levels(annotation_row$Phylum))
getPalette <- colorRampPalette(brewer.pal(colourCount, "Set3"))

#apply to phylum
phylum_col <- getPalette(colourCount)
phylum_col
phylum_col <- c("#8DD3C7",  "#FFFFB3",  "#FB8072", "#BEBADA","#80B1D3")
levels(annotation_row$Phylum)
names(phylum_col) = levels(annotation_row$Phylum)

#apply to guild
colourCount <- length(levels(annotation_row$Guild))
getPalette <- colorRampPalette(brewer.pal(colourCount, "Paired"))
getPalette(colourCount)
guild_colors <- c("#A6CEE3", "#1F78B4", "#B2DF8A", "#33A02C", "#FB9A99", "#E31A1C", "#FDBF6F")
levels(annotation_row$Guild)
guild_colors <- c("gray","#E31A1C", "#FDBF6F", "#A6CEE3", "#1F78B4", "#B2DF8A", "#33A02C")
names(guild_colors) <- levels(annotation_row$Guild)

#change name of experiment so it works
color_experiment <- c("#CC3399", "#70AD47")
names(color_experiment) = c("cond_phytohormones", "real_insects")

#put all colors together
annotation_colors = list(
     Phylum = phylum_col,
     Guild = guild_colors,
     JAvsCtrl = c("TRUE" = "#E69F00", "FALSE" = "white"),
     SAvsCtrl = c("TRUE" = "#4472C4", "FALSE" = "white"),
     Experiment = color_experiment)


#make colors for matrix
#Select colors for the ends of the values and the middle 
function_color_matrix <- circlize::colorRamp2(c(-2.5, 0, 2.5),
                          c("#0088c9",  "white", "#c85200"))

function_color_matrix(seq(-3,3)) #gives you 9 colors with white in the middle 

# now plot the pheatmap
lfc_heatmap <- ComplexHeatmap::pheatmap(matrix.log.EPG_ordered,
                         color= function_color_matrix,
                         name="LFC",
                         scale= "none", 
                         row_labels = species_ordered$Genus,
                         #annotation_col = annotation_col, 
                         annotation_row = annotation_row, 
                         annotation_colors = annotation_colors,
                         cluster_cols = FALSE,
                         cluster_rows = TRUE,
                         main = "EPG-ITS",
                         fontsize_row = 8,
                         fontsize_col = 8)
#view
lfc_heatmap

#save by hand with dimensions: 784x634 px 

```

###Save data
```{r}

ancom.e <- ancom$EPG #ancom results
lfc  #lfc values  
asvs <- sig_asvs$EPG$ASV #vector with ASV names 
dif_abun_stress #contains whether TRUE or FALSE for JA and SA 
log_corr_abn <- log_corr_abn_sig$EPG #corrected abundances of DA ASVs
ps_DA #ps with info

#make it into a list  
epg_dataDA <- list (ancom_test=ancom.e, lfc=lfc, sig_asvs=asvs,
                    binary_stress=dif_abun_stress, log_corr_abn=log_corr_abn,
                    ps=ps_DA)

#save it
save(epg_dataDA, file="./Results/03_differential_abundance/EPG_dataDA_ITS.Rdata")

#clean
rm.all.but(keep=c("ps_ITS_split"), keep_functions = F)
gc()

```

##iTOL

```{r pruning ps}

#load full ITS ps, you'll need it to prune the tree from there 
load("./amplicon_data/Phyloseq_objects/ps_ITS.RData")

#load phyto data
load("./Results/03_differential_abundance/phyto_dataDA.Rdata")

#load insects data
load("./Results/03_differential_abundance/insects_dataDA.Rdata")


#check name of ASVs to prune a ps 
p <- rownames(phyto_dataDA$ps@tax_table)
i <- rownames(insects_dataDA$ps@tax_table)

intersect <- intersect(p,i) #yes, in 3 of them
asvs <- c(p,i)
asvs <- unique(asvs) #58

#First, let's get the tree by pruning the original ps_ITS with DA ASVs
#make ps with DA asvs
ps_DA <- prune_taxa(asvs, ps_ITS) #58 taxa, ok!
ITS_tree <- ps_DA@phy_tree
plot(ITS_tree)
ITS_tree$tip.label

#save it
ape::write.tree(ITS_tree, "./Results/03_differential_abundance/ITS_iTOL/ITS_DA_tree.tre")

#Now, let's get the sequences to make a new tree as the default one with DADA2 is not great
#get the sequences and save them as FASTA format
ps_DA %>%
refseq() %>%
Biostrings::writeXStringSet("./Results/03_differential_abundance/ITS_iTOL/selected_asvs.fna",
             append=FALSE,compress=FALSE,
             compression_level=NA, format="fasta")

#Run alignment in MAFFT, remove noisy columns with Guidance2
#and then use this as an input for IQTREE in the webserver.
#Guidance2 link: https://taux.evolseq.net/guidance/results/17244274565078#remove_pos 23/08

#4. Load phylogenetic tree 
new_tree <- read.tree("./Results/03_differential_abundance/ITS_iTOL/ITS_DA_tree_IQTREE.tre")
new_tree <- ape::multi2di(new_tree) #root it

#check
new_tree
plot(new_tree)

#change tree in the ps object
#copy ps 
ps_DA_tree <- ps_DA

#change it for new
ps_DA_tree@phy_tree <- new_tree

#compare it
ps_DA@phy_tree
ps_DA_tree@phy_tree  

#save it
ape::write.tree(new_tree, "./Results/03_differential_abundance/ITS_iTOL/ITSS_DA_tree_IQTREE_rooted.tre")

#call it again
ps_DA <- ps_DA_tree

#clean
rm(ps_DA_tree)

```

Now, create a complete dataframe with all the info

```{r dataframe}

##let's get D1:Experiment
df.p <- as.data.frame(p) %>% 
         mutate(Phytohormones= "phytohormones")%>%
         mutate(Experiment="phytohormones")  
colnames(df.p)[1] <-  "id"
  
df.i <- as.data.frame(i)%>% 
         mutate(Insects= "insects")%>%
         mutate(Experiment="insects") 
colnames(df.i)[1] <-  "id"

#merge
df_experiment <- full_join(df.p, df.i)

#get order
order_asvs <- df_experiment$id

#Now, let's get D2:LFC
#get LFC values from both experiments in a single df 
lfc_phyto <- phyto_dataDA$lfc %>%  mutate(Phytohormones= "phytohormones") %>% 
                                   rownames_to_column(var="id")
colnames(lfc_phyto) <- c("id", "lfc_JA_phyto", "lfc_SA_phyto", "Phytohormones")                              

lfc_insects <- insects_dataDA$lfc %>%  mutate(Insects= "insects") %>% 
                                   rownames_to_column(var="id")
colnames(lfc_insects) <- c("id", "lfc_JA_insects", "lfc_SA_insects", "Insects")  

#merge
df_lfc <- full_join(lfc_phyto, lfc_insects)

#Next, let's get D3:Treatment
#phyto
bin_phyto <- phyto_dataDA$binary_stress %>% rownames_to_column(var="id") %>% mutate(Phytohormones="phytohormones")
colnames(bin_phyto) <- c("id","DA_JA_phyto", "DA_SA_phyto", "Phytohormones")
bin_phyto <- bin_phyto %>% mutate(DA_JA_phyto = case_when(DA_JA_phyto == TRUE ~ "YES",
                                DA_JA_phyto == FALSE ~ "NO")) %>% 
                              mutate(DA_SA_phyto = case_when(DA_SA_phyto == TRUE ~ "YES",
                                DA_SA_phyto == FALSE ~ "NO"))
bin_phyto <- bin_phyto %>%
                mutate(DA_phyto = case_when(
                  DA_JA_phyto == "YES" & DA_SA_phyto == "NO" ~ "JA_Only",
                  DA_JA_phyto == "NO" & DA_SA_phyto == "YES" ~ "SA_Only",
                  DA_JA_phyto == "YES" & DA_SA_phyto == "YES" ~ "JA_and_SA")) #works

#insects
bin_insects <- insects_dataDA$binary_stress %>% rownames_to_column(var="id") %>% mutate(Insects="insects")
colnames(bin_insects) <- c("id","DA_JA_insects", "DA_SA_insects", "Insects")
bin_insects <- bin_insects %>% mutate(DA_JA_insects = case_when(DA_JA_insects == TRUE ~ "YES",
                                DA_JA_insects == FALSE ~ "NO")) %>% 
                              mutate(DA_SA_insects = case_when(DA_SA_insects == TRUE ~ "YES",
                                DA_SA_insects == FALSE ~ "NO"))
bin_insects <- bin_insects %>%
                mutate(DA_insects = case_when(
                  DA_JA_insects == "YES" & DA_SA_insects == "NO" ~ "JA_Only",
                  DA_JA_insects == "NO" & DA_SA_insects == "YES" ~ "SA_Only",
                  DA_JA_insects == "YES" & DA_SA_insects == "YES" ~ "JA_and_SA")) #works
  
#merge
binary_stress <- full_join(bin_phyto, bin_insects)

#merge to make it complete
df_complete <- full_join(df_lfc, binary_stress)

#add column with Experiment including Phyto, Insects and both
df_complete <- df_complete %>% mutate(Experiment=case_when(
  Phytohormones == "phytohormones" ~ "phytohormones",
  Insects == "insects" ~ "insects"))

#re-order
df_complete <- df_complete %>% relocate(id, Experiment, Phytohormones, Insects)

```


```{r dataframe-taxonomy}

#Finally, let's get taxonomy
#clean it first
tax_table(ps_DA)[, colnames(tax_table(ps_DA))] <- gsub(tax_table(ps_DA)[, colnames(tax_table(ps_DA))],     pattern = "[a-z]__", replacement = "")
#make it a df 
tax_table <- as.data.frame(ps_DA@tax_table[,c(1:7,12)]) #select the interesting ones 
tax_table$id <- rownames(tax_table) #add id column for join 

#merge
df_complete2 <- left_join(df_complete, tax_table)

#Now this df is complete (for now)
#save it
saveRDS(df_complete2, file="./Data_frames/03_differential_abundance/df_complete_ITS.Rdata")
write.csv(df_complete2, file="./Data_frames/03_differential_abundance/df_complete_ITS.csv")

```

Move it into iTOL format

iTOL has a specific format to add data into the tree, it has to be in a .txt file and in a certain way coded. The package itol.toolkit helps to make an annotation file directly in R.


```{r color management}

#color management

#Set-up colors
color_JA <- "#E69F00"
color_SA <- "#4472C4"
color_phyto <- "#CC3399"
color_insects <- "#70AD47"

#Built in colors
display.brewer.all() #gives you the palettes from RColorBrewer


```

```{r Experiment}

#create Annotation files with using the itol.toolkit package

#Experiment: together
temp <- df_complete2%>%
        dplyr::select(c(id,Experiment))

#make df for itol
AF1 <- itol.toolkit::create_unit(data = temp, 
                         key = "AF1_Experiment", 
                         type = "DATASET_COLORSTRIP",
                         tree = ITS_tree)

#change colors and parameters
AF1@specific_themes$strip_label$width <- 30 #width of strip
AF1@specific_themes$strip_label$color_branches <- 0 #no color branches
AF1@data$tip <- AF1@data$tip %>%  mutate(`AF1_Experiment$COLOR`=case_when( #change colors 
                                `AF1_Experiment$LABEL`=="phytohormones" ~ color_phyto,
                                `AF1_Experiment$LABEL`=="insects" ~ color_insects))


AF1@common_themes$legend$title <- "Experiment"
AF1@common_themes$legend$horizontal <- 0
AF1@common_themes$legend$labels <- c("Herbivore-induced","Phytohormone-induced")

#save
itol.toolkit::write_unit(AF1, "./Results/03_differential_abundance/ITS_iTOL") 
```

Treatment
```{r Treatment}
## Treatment
df_Treatment <- df_complete2%>%
                dplyr::select(id,DA_JA_phyto,DA_JA_insects, DA_SA_phyto, DA_SA_insects)%>%
                relocate(id)
#replace NA's
df_Treatment$DA_JA_phyto <- df_Treatment$DA_JA_phyto %>% replace_na('NO')
df_Treatment$DA_SA_phyto <- df_Treatment$DA_SA_phyto %>% replace_na('NO')
df_Treatment$DA_JA_insects <- df_Treatment$DA_JA_insects %>% replace_na('NO')
df_Treatment$DA_SA_insects <- df_Treatment$DA_SA_insects %>% replace_na('NO')

#replace 0's for -1 and Yes for 1 so iTOL recognizes full and empty symbols
df_Treatment2 <- df_Treatment %>% 
                mutate(DA_JA_phyto=case_when(DA_JA_phyto== "YES" ~ "1", TRUE ~ "-1")) %>% 
                mutate(DA_SA_phyto=case_when(DA_SA_phyto== "YES" ~ "1", TRUE ~ "-1")) %>%
                mutate(DA_JA_insects=case_when(DA_JA_insects== "YES" ~ "1", TRUE ~ "-1")) %>%
                mutate(DA_SA_insects=case_when(DA_SA_insects== "YES" ~ "1", TRUE ~ "-1"))

## By Experiment
#Phyto
Phyto <- df_Treatment2 %>% dplyr::select(id, DA_JA_phyto, DA_SA_phyto)
AF2.1 <- itol.toolkit::create_unit(data = Phyto, 
                         key = "AF2.1_Treatment_Phyto", 
                         type = "DATASET_BINARY",
                         tree = ITS_tree)

AF2.1@field$labels <- c("JA_phyto",  "SA_phyto")
AF2.1@field$colors <- c("#E69F00",  "#4472C4")

itol.toolkit::write_unit(AF2.1, "./Results/03_differential_abundance/ITS_iTOL")

#Insects
insects <- df_Treatment2 %>% dplyr::select(id, DA_JA_insects, DA_SA_insects)
AF2.2 <- itol.toolkit::create_unit(data = insects, 
                         key = "AF2.2_Treatment_insects", 
                         type = "DATASET_BINARY",
                         tree = ITS_tree)

AF2.2@field$labels <- c("JA_insects",  "SA_insects")
AF2.2@field$colors <- c("#E69F00",  "#4472C4")

itol.toolkit::write_unit(AF2.2, "./Results/03_differential_abundance/ITS_iTOL")

```

LFC
```{r LFC-heatmap}

## LFC dataset
df_LFC <- df_complete2%>%
                dplyr::select(id,lfc_JA_phyto, lfc_JA_insects, lfc_SA_phyto, lfc_SA_insects)%>%
                relocate(id)%>%
                as.tibble()
#let's round up
df_LFC <- df_LFC %>% 
          dplyr::mutate(across(where(is.numeric), round, 2))

#First, lets make a single df 
df_LFC_unique <- full_join(df_Treatment,df_LFC)
#Now let's apply some rules
df_LFC_unique <- df_LFC_unique %>%
                  mutate(lfc_JA_phyto=
                           case_when(DA_JA_phyto == "NO" ~ NA, TRUE ~ lfc_JA_phyto))%>% 
                  mutate(lfc_SA_phyto=
                           case_when(DA_SA_phyto == "NO" ~ NA, TRUE ~ lfc_SA_phyto))%>% 
                  mutate(lfc_JA_insects=
                           case_when(DA_JA_insects == "NO" ~ NA, TRUE ~ lfc_JA_insects))%>% 
                  mutate(lfc_SA_insects=
                           case_when(DA_SA_insects == "NO" ~ NA, TRUE ~ lfc_SA_insects))

#By experiment
Phyto <-  df_LFC_unique %>% dplyr::select(id, lfc_JA_phyto, lfc_SA_phyto)

#get Annotation File for Heatmap
AF3.1 <- itol.toolkit::create_unit(data = Phyto , 
                         key = "AF3.1_LFC_Heatmap_Phyto", 
                         type = "DATASET_HEATMAP",
                         tree = ITS_tree)  #works

AF3.1@data$tip[AF3.1@data$tip == 0] <- NA

#choose colors
display.brewer.pal(n=8, name='RdBu') #see palette
color_heatmap <- brewer.pal(n = 8, name = "RdBu") #get colors
min <- color_heatmap[8]
max <- color_heatmap[1]

#select colors 
AF3.1@specific_themes$heatmap$color$min <- min
AF3.1@specific_themes$heatmap$color$max <- max
AF3.1@specific_themes$heatmap$color$mid <- "#ebeceb"
AF3.1@specific_themes$heatmap$value$mid <- 0
AF3.1@specific_themes$heatmap$color$nan <- "#FFFFFF"

#change range so the colors mean the same across experiment and treatment
AF3.1@specific_themes$heatmap$value$max <- 2.89
AF3.1@specific_themes$heatmap$value$min <- -2.61
AF3.1@specific_themes$heatmap$value$mid <- 0

AF3.1@specific_themes$heatmap$tree$tree_display <- 0 #no display of tree in heatmap

itol.toolkit::write_unit(AF3.1, "./Results/03_differential_abundance/ITS_iTOL") #works

#By experiment
Insects <-  df_LFC_unique %>% dplyr::select(id, lfc_JA_insects, lfc_SA_insects)

#get Annotation File for Heatmap
AF3.2 <- itol.toolkit::create_unit(data = Insects , 
                         key = "AF3.2_LFC_Heatmap_Insects", 
                         type = "DATASET_HEATMAP",
                         tree = ITS_tree)  #works

AF3.2@data$tip[AF3.2@data$tip == 0] <- NA

#select colors 
AF3.2@specific_themes$heatmap$color$min <- min
AF3.2@specific_themes$heatmap$color$max <- max
AF3.2@specific_themes$heatmap$color$mid <- "#ebeceb"
AF3.2@specific_themes$heatmap$value$mid <- 0
AF3.2@specific_themes$heatmap$color$nan <- "#FFFFFF"

#change range so the colors mean the same across experiment and treatment
AF3.2@specific_themes$heatmap$value$max <- 2.89
AF3.2@specific_themes$heatmap$value$min <- -2.61
AF3.2@specific_themes$heatmap$value$mid <- 0

AF3.2@specific_themes$heatmap$tree$tree_display <- 0 #no display of tree in heatmap

itol.toolkit::write_unit(AF3.2, "./Results/03_differential_abundance/ITS_iTOL") #works

```

Now I need to add the taxonomy and change the labels

```{r Taxonomy}

#first, let's change it for Family labels
#Family
df_labels_F <- df_complete2 %>%
               dplyr::select(id, Family)

AF4 <- create_unit(data = df_labels_F,
                       key = "AF4_label_Family",
                       type = "LABELS",
                       tree = ITS_tree)

itol.toolkit::write_unit(AF4, "./Results/03_differential_abundance/ITS_iTOL")

#Genus
df_labels_G <- df_complete2 %>%
               dplyr::select(id, Genus)

#replace Genus names with names of ASV and add number to Genus so its clear that it's different ASVs
df_labels_G <- df_labels_G %>% mutate(number=str_split_i(df_labels_G$id, "_", 2))%>% #get number of ASV
                                mutate(Genus=paste(.$Genus, .$number, sep= "-"))%>% #make new with number and genus
                                mutate(Genus=gsub("NA", "fASV", .$Genus))%>%
                                dplyr::select(-number) #ok
                                      
AF5 <- create_unit(data = df_labels_G,
                       key = "AF5_label_Genus",
                       type = "LABELS",
                       tree = ITS_tree)

itol.toolkit::write_unit(AF5, "./Results/03_differential_abundance/ITS_iTOL")


#Make a range df for Phylum now
df_Phylum <- df_complete2 %>%
               dplyr::select(id, Phylum)

df_Phylum$Phylum <- df_Phylum$Phylum %>% replace_na('Unasigned')

#make a vector with colors 
Phylum <- unique(df_Phylum$Phylum) #5
#colourCount <- 5
#getPalette <- colorRampPalette(brewer.pal(colourCount, "Pastel1"))
#color_phylum <- getPalette(colourCount)
#color_phylum <- c("#FBB4AE", "#B3CDE3", "#CCEBC5", "#DECBE4", "#FED9A6")
color_phylum <- c("#FDB462","#BC80BD","#FFFFB3", "#FB8072","#80B1D3")

#make it a df
df_phyla_col <- data.frame(Phylum, color_phylum)

#merge with df_Phylum
df_Phylum <- df_Phylum%>%
             dplyr::inner_join(df_phyla_col)

AF6 <- create_unit(data = df_Phylum, 
                      key = "AF6_Phylum_colors", 
                      type = "TREE_COLORS", 
                      subtype = "range",
                      tree = ITS_tree)

itol.toolkit::write_unit(AF6, "./Results/03_differential_abundance/ITS_iTOL")
```


```{r trophic-mode }

#Finally, make a strip df for life-style
df_trophic <- df_complete2 %>%
               dplyr::select(id, Trophic.Mode)

AF7 <- itol.toolkit::create_unit(data = df_trophic, 
                         key = "AF7_Trophic_mode", 
                         type = "DATASET_SYMBOL",
                         tree = ITS_tree)

AF7@data$tip$`AF10_Trophic_mode$POSITION`<- 1
AF7@data$tip$`AF7_Trophic_mode$SYMBOL` <- 4


itol.toolkit::write_unit(AF7, "./Results/03_differential_abundance/ITS_iTOL")

```


## Raw abundances
Quickly plot the raw abundances of those DA ASVs to check them

```{r}
#temp
otu_table <- as.data.frame(otu_table(ps_DA))
otu_table <- otu_table %>% mutate(id=rownames(.))

df_complete3 <- left_join(df_complete2, otu_table, by="id")
str(df_complete3)

df_complete4[c(1:2,4:12)] <- lapply(df_complete4[c(1:2,4:12)], factor)

#make it long
long <- df_complete3 %>% 
              tidyr::pivot_longer(cols = c(22:145), 
              names_to = "sample",
              values_to = "raw_counts",
              values_drop_na = TRUE)

#almost ready, I just need a column with plant_pathway_induced based on sample
df_samples <- data.frame(sample_data(ps_ITS))
df_samples <- df_samples %>% mutate(sample=rownames(.))%>%
              dplyr::select(c('sample', 'plant_pathway_induced')) %>% 
              mutate(plant_pathway_induced=factor(plant_pathway_induced))

levels(df_samples$plant_pathway_induced) <- c("Control", "Control", "JA" ,"SA")    

#now I just need to add this with some join
long_complete <- dplyr::left_join(long, df_samples, by="sample")
str(long_complete)
long_complete[c(1,4,7:22)] <- lapply(long_complete[c(1,4,7:22)], factor)

# Set colors
#Set colors
color_pathway_induced <- c("#BFBFBF", "#E69F00", "#4472C4")
color_experiment <- c("#CC3399", "#70AD47")
color_ctrl_vs_JA <- c("#BFBFBF", "#E69F00")
color_ctrl_vs_SA <- c("#BFBFBF", "#4472C4")


plot <- ggplot(data=subset(long_complete, Experiment=="phytohormones" &
                           DA_phyto=="JA_and_SA"),
               aes(x=plant_pathway_induced, y=raw_counts), fill=plant_pathway_induced)+
        geom_violin(alpha=0.9, colour="black", linewidth=0.6, aes(fill=plant_pathway_induced))+
        scale_fill_manual(values = color_pathway_induced)+
        facet_wrap(id~Experiment, scales="free", ncol=6)+
        geom_jitter(position = position_jitter(w = 0.1, h = 0),
                        size=1, shape=20, alpha=0.8, aes(fill=plant_pathway_induced))

levels(long_complete$DA_phyto)
ggplot(data=subset(long_complete, Experiment=="insects" &
                           DA_insects=="SA_Only"),
               aes(x=plant_pathway_induced, y=raw_counts), fill=plant_pathway_induced)+
        geom_violin(alpha=0.9, colour="black", linewidth=0.6, aes(fill=plant_pathway_induced))+
        scale_fill_manual(values = color_pathway_induced)+
        facet_wrap(id~Experiment, scales="free", ncol=6)+
        geom_jitter(position = position_jitter(w = 0.1, h = 0),
                        size=1, shape=20, alpha=0.8, aes(fill=plant_pathway_induced))


```


